{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Variant calling analysis Killswitch circuit \n",
    "\n",
    "Defining variants (SNPs, insertions and deletions) for a set of HiSeq samples exploring the variability in populations of *M. pneumoniae* transformed with a killswitch cassette (C5 samples) or two (C20 samples); sequencing performed at different selection passages (2, 3, 15; 2 replicates in each condition and cassette) and a pair of replicates at passage 3 were the cassette was induced by IPTG.\n",
    "\n",
    "**Authors:** [Samuel Miravet-Verde](mailto:samuel.miravet@crg.eu) and [Alicia Broto](mailto:alicia.broto@crg.eu) \n",
    "\n",
    "**Last update:** 14/06/2021"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib widget\n",
    "\n",
    "import sys, os\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scripts\n",
    "from collections import Counter\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "sns.set_style(\"whitegrid\", {'axes.grid' : False})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# General info about the genome references and the location of the cassettes in the genomes\n",
    "seq_record5 = SeqIO.read('./data/C5_refSeq.gb', \"genbank\")\n",
    "seq_record20 = SeqIO.read('./data/C20_refSeq.gb', \"genbank\")\n",
    "genome_C5 = str(seq_record5.seq)\n",
    "genome_C20 = str(seq_record5.seq)\n",
    "\n",
    "cass_par = [565510, 572124] # Present in C5 and C20\n",
    "cass_3b = [780850, 787445]  # Only in C20\n",
    "cass_26 = [574473, 581077]\n",
    "cass_262 = [79542, 88496]\n",
    "\n",
    "\n",
    "# Annotation information to work faster with dataframes\n",
    "ncbi = pd.read_csv('./data/mpn_annotation.csv', sep='\\t', header=None) # Genome annotation\n",
    "ncbi.columns = ['gene', 'start', 'end', 'strand']\n",
    "ncbi.set_index('gene', inplace=True)\n",
    "gold = pd.read_csv('./data/goldsets.csv', sep='\\t')  # Gold set \n",
    "\n",
    "posN = []\n",
    "posE = []\n",
    "for gene, cat in zip(gold['gene'], gold['class']):\n",
    "    if cat=='E':\n",
    "        posE += range(ncbi.loc[gene][0], ncbi.loc[gene][1]+1)\n",
    "    else:\n",
    "        posN += range(ncbi.loc[gene][0], ncbi.loc[gene][1]+1)\n",
    "        \n",
    "# Define the different set of positions in the genome were we will perform the analysis\n",
    "posN = set(posN)\n",
    "posE = set(posE)\n",
    "cas1 = set(range(cass_par[0], cass_par[1]+1))\n",
    "cas2 = set(range(cass_3b[0], cass_3b[1]+1))\n",
    "cas12 = cas1.union(cas2)\n",
    "cas31 = set(range(cass_26[0], cass_26[1]+1))\n",
    "cas32 = set(range(cass_262[0], cass_262[1]+1))\n",
    "cas33 = cas31.union(cas32)\n",
    "\n",
    "geno = set(range(1, 816395))\n",
    "\n",
    "ingene = []\n",
    "for st, en in zip(ncbi['start'], ncbi['end']):\n",
    "    ingene+=list(range(st, en))\n",
    "ingene = set(ingene)\n",
    "outgene = geno.difference(ingene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Map variations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will be using [Snippy](https://github.com/tseemann/snippy) for the first steps related to sequence mapping. To run it in parallel for all the samples we create a input file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/C20p3/B_S1_L001_R1_001.fastq.gz',\n",
       " '../data/C20p3/B_S2_L001_R1_001.fastq.gz',\n",
       " '../data/C5p3/D_S2_L001_R1_001.fastq.gz',\n",
       " '../data/C5p3/D_S3_L001_R1_001.fastq.gz',\n",
       " '../data/C5p3IPTG/A_S1_L001_R1_001.fastq.gz',\n",
       " '../data/C5p3IPTG/A_S1b_L001_R1_001.fastq.gz',\n",
       " '../data/C20p15/7_S3_L001_R1_001.fastq.gz',\n",
       " '../data/C20p15/7_S5_L001_R1_001.fastq.gz',\n",
       " '../data/C5p15/F_S5_L001_R1_001.fastq.gz',\n",
       " '../data/C5p15/F_S1_L001_R1_001.fastq.gz',\n",
       " '../data/C5p2/2_S2_L001_R1_001.fastq.gz',\n",
       " '../data/C5p2/2_S3_L001_R1_001.fastq.gz',\n",
       " '../data/C5p2/2_S1_L001_R1_001.fastq.gz',\n",
       " '../data/C26p2/8_S2_L001_R1_001.fastq.gz',\n",
       " '../data/C26p2/8_S3_L001_R1_001.fastq.gz',\n",
       " '../data/C20p2/6_S2_L001_R1_001.fastq.gz',\n",
       " '../data/C20p2/6_S4_L001_R1_001.fastq.gz',\n",
       " '../data/C26p15/9_S7_L001_R1_001.fastq.gz',\n",
       " '../data/C26p15/9_S3_L001_R1_001.fastq.gz']"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### 1. Make a multifile to map with snippy-multi.\n",
    "files = glob.glob('../data/C[5|20|26]*/*_R1_*') # The gre expression refers to the raw sequencing files located in data (this has to point to the directory where fastq files are stored)\n",
    "\n",
    "fo1 = open('./tmp_files/snippy_input1.txt', 'w')\n",
    "fo2 = open('./tmp_files/snippy_input2.txt', 'w')\n",
    "fo3 = open('./tmp_files/snippy_input3.txt', 'w')\n",
    "for f in files:\n",
    "    ide = f.split('/')[2:]\n",
    "    if 'C5' in f:\n",
    "        ide = ide[0].replace('C5', '')+'_'+'_'.join(ide[1].split('_')[:-3])\n",
    "        fo1.write('{}\\t{}\\t{}\\n'.format(ide, f, f.replace('_R1_', '_R2_')))\n",
    "    elif 'C20' in f:\n",
    "        ide = ide[0].replace('C20', '')+'_'+'_'.join(ide[1].split('_')[:-3])\n",
    "        fo2.write('{}\\t{}\\t{}\\n'.format(ide, f, f.replace('_R1_', '_R2_')))\n",
    "    else:\n",
    "        ide = ide[0].replace('C26', '')+'_'+'_'.join(ide[1].split('_')[:-3])\n",
    "        fo3.write('{}\\t{}\\t{}\\n'.format(ide, f, f.replace('_R1_', '_R2_')))        \n",
    "fo1.close()\n",
    "fo2.close()\n",
    "fo3.close()\n",
    "files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now run the command in the snippy environment:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "conda activate snippy\n",
    "snippy-multi ./tmp_files/snippy_input.txt --ref ./data/C5_refSeq.gb --cpus 4 --report >  ./snippy_run.sh\n",
    "sh ./snippy_run.sh\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Listed samples in order\n",
    "order1 = ['p2_2_S1','p2_2_S2','p2_2_S3','p3_D_S2', 'p3_D_S3','p15_F_S1','p15_F_S5', 'p3IPTG_A_S1', 'p3IPTG_A_S1b']\n",
    "order2 = ['p2_6_S2','p2_6_S4','p3_B_S1','p3_B_S2','p15_7_S3','p15_7_S5']\n",
    "order3 = ['p2_8_S2','p2_8_S3','p15_9_S3','p15_9_S7']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "freebayes-parallel p2_8_S2/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p2_8_S2/reference/ref.fa ../p2_8_S2/snps.bam > filter3/p2_8_S2.raw.vcf\n",
      "freebayes-parallel p2_8_S3/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p2_8_S3/reference/ref.fa ../p2_8_S3/snps.bam > filter3/p2_8_S3.raw.vcf\n",
      "freebayes-parallel p15_9_S3/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p15_9_S3/reference/ref.fa ../p15_9_S3/snps.bam > filter3/p15_9_S3.raw.vcf\n",
      "freebayes-parallel p15_9_S7/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f p15_9_S7/reference/ref.fa ../p15_9_S7/snps.bam > filter3/p15_9_S7.raw.vcf\n"
     ]
    }
   ],
   "source": [
    "# freebayes-parallel reference/ref.txt 4 -p 2 -P 0 -C 2 -F 0.05 --min-coverage 10 --min-repeat-entropy 1.0 -q 13 -m 60 --strict-vcf   -f reference/ref.fa snps.bam > snps.raw.vcf\n",
    "for fil in glob.glob('../p*/*.bam'):\n",
    "    # again this has to point to directories were snipy output is stored\n",
    "    ide = fil.split('/')[1]\n",
    "    #print('freebayes-parallel {}/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.01 -q 13 -m 60 --strict-vcf -f {}/reference/ref.fa {} > filter10/{}.raw.vcf'.format(ide, ide, fil, ide))\n",
    "    #print('freebayes-parallel {}/reference/ref.txt 4 -p 1 -P 0 -C 2 -F 0.05 -q 13 -m 60 --strict-vcf -f {}/reference/ref.fa {} > filter50/{}.raw.vcf'.format(ide, ide, fil, ide))\n",
    "    if ide in order1:\n",
    "        pass\n",
    "        print('freebayes-parallel {}/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f {}/reference/ref.fa {} > filter1/{}.raw.vcf'.format(ide, ide, fil, ide))\n",
    "    elif ide in order2:\n",
    "        pass\n",
    "        print('freebayes-parallel {}/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f {}/reference/ref.fa {} > filter2/{}.raw.vcf'.format(ide, ide, fil, ide))\n",
    "    else:\n",
    "        print('freebayes-parallel {}/reference/ref.txt 4 -p 1 -P 0 -C 1 -F 0.001 -q 13 -m 60 --strict-vcf -f {}/reference/ref.fa {} > filter3/{}.raw.vcf'.format(ide, ide, fil, ide))\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get the total number of mapped reads:\n",
    "\n",
    "```python\n",
    "for fil in glob.glob('./p*/*.bam'):\n",
    "    ide = fil.split('/')[1]\n",
    "    # get the total number of reads of a BAM file (may include unmapped and duplicated multi-aligned reads)\n",
    "    print('samtools view -c {} >> total_readcount.txt'.format(fil))\n",
    "    print('samtools view -c {} -F 260 >> total_readcount_mapped.txt'.format(fil))\n",
    "```\n",
    "\n",
    "Finally we annotate the effect of the variants using SnpEff. This requires to add the custom annotation to the sources, specifying that the translation code is set to 4 for *M. pneumoniae*.\n",
    "\n",
    "1. Download SnpEff and edited the config (in software_crg/SnpEff) to include the C5, C20, C26 refseq references\n",
    "2. Build the custom db:\n",
    "```bash\n",
    "java -jar snpEff.jar build -genbank -v C5ali\n",
    "java -jar snpEff.jar build -genbank -v C20ali\n",
    "java -jar snpEff.jar build -genbank -v C26ali\n",
    "```\n",
    "3. Annotate vcf files in bash (in directory of the software)\n",
    "```bash\n",
    "for fil in `ls ./filter1/p*.vcf`; do java -Xmx8g -jar snpEff.jar C5ali $fil > $fil.ann; done\n",
    "for fil in `ls ./filter2/p*.vcf`; do java -Xmx8g -jar snpEff.jar C20ali $fil > $fil.ann; done\n",
    "for fil in `ls ./filter3/p*.vcf`; do java -Xmx8g -jar snpEff.jar C26ali $fil > $fil.ann; done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## 2. Data loading \n",
    "\n",
    "We will parse the variations to keep the columns of interest to perform the exploratory analysis. Dataframes include the following information (by column header):\n",
    "- SAMPLE: sample identifier includes information about the passage (e.g. p3IPTG is the passage 3 induced)\n",
    "- PASS: passage (2,3 or 15)\n",
    "- POS: genome loci (base pair position) where the variant is found\n",
    "- QUAL: estimate of the probability that there is a polymorphism at the loci described by the record\n",
    "- TOT: total number of reads covering a loci\n",
    "- REFN: total number of reads covering a loci matching the reference\n",
    "- ALTN: number of reads presenting a variant\n",
    "- REF: reference sequence\n",
    "- ALT: alternative sequence\n",
    "- EFF: potential effect of the variant as given by snpeff\n",
    "- IMPACT: potential impact of the variant, can be LOW (synonimous mutations), MODERATE (missense), HIGH (non-synonimous mutations, start lost, stop lost), or MODIFIER (when it occurs in an intergenic region)\n",
    "- AFF: gene affected\n",
    "- MUT: mutation (in nucleotide if not coding in amino acid if coding)\n",
    "- ANN_TYPE: annotation of the loci, can be intergenic, gene, or cassette (covering the principal elements in the cassette)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing data for genome 5\n",
      "Cleaning data for genome 5\n",
      "Memory usage of properties dataframe is : 62.0408935546875  MB\n",
      "******************************\n",
      "Column:  PASS\n",
      "dtype before:  int64\n",
      "dtype after:  uint8\n",
      "******************************\n",
      "******************************\n",
      "Column:  POS\n",
      "dtype before:  int64\n",
      "dtype after:  uint32\n",
      "******************************\n",
      "******************************\n",
      "Column:  QUAL\n",
      "dtype before:  float64\n",
      "dtype after:  float32\n",
      "******************************\n",
      "******************************\n",
      "Column:  TOT\n",
      "dtype before:  int64\n",
      "dtype after:  uint16\n",
      "******************************\n",
      "******************************\n",
      "Column:  REFN\n",
      "dtype before:  int64\n",
      "dtype after:  uint16\n",
      "******************************\n",
      "******************************\n",
      "Column:  ALTN\n",
      "dtype before:  int64\n",
      "dtype after:  uint8\n",
      "******************************\n",
      "******************************\n",
      "Column:  FRAC\n",
      "dtype before:  float64\n",
      "dtype after:  float32\n",
      "******************************\n",
      "___MEMORY USAGE AFTER COMPLETION:___\n",
      "Memory usage is:  43.62250328063965  MB\n",
      "This is  70.3125 % of the initial size\n",
      "Saving data for genome 5\n",
      "Processing data for genome 20\n",
      "Cleaning data for genome 20\n",
      "Memory usage of properties dataframe is : 51.05322265625  MB\n",
      "******************************\n",
      "Column:  PASS\n",
      "dtype before:  int64\n",
      "dtype after:  uint8\n",
      "******************************\n",
      "******************************\n",
      "Column:  POS\n",
      "dtype before:  int64\n",
      "dtype after:  uint32\n",
      "******************************\n",
      "******************************\n",
      "Column:  QUAL\n",
      "dtype before:  float64\n",
      "dtype after:  float32\n",
      "******************************\n",
      "******************************\n",
      "Column:  TOT\n",
      "dtype before:  int64\n",
      "dtype after:  uint16\n",
      "******************************\n",
      "******************************\n",
      "Column:  REFN\n",
      "dtype before:  int64\n",
      "dtype after:  uint16\n",
      "******************************\n",
      "******************************\n",
      "Column:  ALTN\n",
      "dtype before:  int64\n",
      "dtype after:  uint8\n",
      "******************************\n",
      "******************************\n",
      "Column:  FRAC\n",
      "dtype before:  float64\n",
      "dtype after:  float32\n",
      "******************************\n",
      "___MEMORY USAGE AFTER COMPLETION:___\n",
      "Memory usage is:  35.89679718017578  MB\n",
      "This is  70.3125 % of the initial size\n",
      "Saving data for genome 20\n",
      "Processing data for genome 26\n",
      "Cleaning data for genome 26\n",
      "Memory usage of properties dataframe is : 36.2364501953125  MB\n",
      "******************************\n",
      "Column:  PASS\n",
      "dtype before:  int64\n",
      "dtype after:  uint8\n",
      "******************************\n",
      "******************************\n",
      "Column:  POS\n",
      "dtype before:  int64\n",
      "dtype after:  uint32\n",
      "******************************\n",
      "******************************\n",
      "Column:  QUAL\n",
      "dtype before:  float64\n",
      "dtype after:  float32\n",
      "******************************\n",
      "******************************\n",
      "Column:  TOT\n",
      "dtype before:  int64\n",
      "dtype after:  uint16\n",
      "******************************\n",
      "******************************\n",
      "Column:  REFN\n",
      "dtype before:  int64\n",
      "dtype after:  uint16\n",
      "******************************\n",
      "******************************\n",
      "Column:  ALTN\n",
      "dtype before:  int64\n",
      "dtype after:  uint16\n",
      "******************************\n",
      "******************************\n",
      "Column:  FRAC\n",
      "dtype before:  float64\n",
      "dtype after:  float32\n",
      "******************************\n",
      "___MEMORY USAGE AFTER COMPLETION:___\n",
      "Memory usage is:  25.76185131072998  MB\n",
      "This is  71.09375 % of the initial size\n",
      "Saving data for genome 26\n"
     ]
    }
   ],
   "source": [
    "# This will concat all the annotated vcf files keeping the fields of interest\n",
    "data1 = scripts.parse_variations('./filter1/*.raw.vcf.ann', order1, genome=5, fname='./results/data1.pickle')\n",
    "data2 = scripts.parse_variations('./filter2/*.raw.vcf.ann', order2, genome=20, fname='./results/data2.pickle')\n",
    "data3 = scripts.parse_variations('./filter3/*.raw.vcf.ann', order3, genome=26, fname='./results/data3.pickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15560"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cas33)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now explore the variants, for example to extract those present in the cassette with high impact (non-synonimous) in samples induced (assigned as passage 18 despite they are passage 3), could be retrieved as:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>443045</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566372</td>\n",
       "      <td>2.184800e-15</td>\n",
       "      <td>677</td>\n",
       "      <td>676</td>\n",
       "      <td>1</td>\n",
       "      <td>0.147929</td>\n",
       "      <td>CTA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuGly102*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443082</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566622</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>647</td>\n",
       "      <td>646</td>\n",
       "      <td>1</td>\n",
       "      <td>0.154799</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Tyr19*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443002</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>565989</td>\n",
       "      <td>2.863080e-15</td>\n",
       "      <td>582</td>\n",
       "      <td>581</td>\n",
       "      <td>1</td>\n",
       "      <td>0.172117</td>\n",
       "      <td>TCCCTCG</td>\n",
       "      <td>TCCTCG</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Gly230fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443294</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>568518</td>\n",
       "      <td>7.700020e-15</td>\n",
       "      <td>577</td>\n",
       "      <td>576</td>\n",
       "      <td>1</td>\n",
       "      <td>0.173611</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Gln1101*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443402</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>569451</td>\n",
       "      <td>6.132760e-15</td>\n",
       "      <td>572</td>\n",
       "      <td>571</td>\n",
       "      <td>1</td>\n",
       "      <td>0.175131</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Glu790*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443663</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>571681</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>481</td>\n",
       "      <td>469</td>\n",
       "      <td>12</td>\n",
       "      <td>2.558635</td>\n",
       "      <td>ATTTTTTTTGATA</td>\n",
       "      <td>ATTTTTTTGATA</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Asn46fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>495218</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>571681</td>\n",
       "      <td>7.895900e-15</td>\n",
       "      <td>191</td>\n",
       "      <td>182</td>\n",
       "      <td>9</td>\n",
       "      <td>4.945055</td>\n",
       "      <td>ATTTTTTTTGATACT</td>\n",
       "      <td>ATTTCTTTAGAAACA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.SerIleLysLys42*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>495059</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>569127</td>\n",
       "      <td>8.076710e-15</td>\n",
       "      <td>235</td>\n",
       "      <td>210</td>\n",
       "      <td>25</td>\n",
       "      <td>11.904762</td>\n",
       "      <td>CAAACT</td>\n",
       "      <td>CAACT</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Phe897fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443366</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>569125</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>534</td>\n",
       "      <td>463</td>\n",
       "      <td>71</td>\n",
       "      <td>15.334773</td>\n",
       "      <td>ATCAAACT</td>\n",
       "      <td>GTCAACC</td>\n",
       "      <td>frameshift_variant&amp;missense_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Lys896fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443009</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566053</td>\n",
       "      <td>1.947940e-13</td>\n",
       "      <td>559</td>\n",
       "      <td>446</td>\n",
       "      <td>113</td>\n",
       "      <td>25.336323</td>\n",
       "      <td>CGCAAA</td>\n",
       "      <td>TGCTAT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuArg208*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN       FRAC  \\\n",
       "443045   p3IPTG_A_S1    18  566372  2.184800e-15  677   676     1   0.147929   \n",
       "443082   p3IPTG_A_S1    18  566622  0.000000e+00  647   646     1   0.154799   \n",
       "443002   p3IPTG_A_S1    18  565989  2.863080e-15  582   581     1   0.172117   \n",
       "443294   p3IPTG_A_S1    18  568518  7.700020e-15  577   576     1   0.173611   \n",
       "443402   p3IPTG_A_S1    18  569451  6.132760e-15  572   571     1   0.175131   \n",
       "...              ...   ...     ...           ...  ...   ...   ...        ...   \n",
       "443663   p3IPTG_A_S1    18  571681  0.000000e+00  481   469    12   2.558635   \n",
       "495218  p3IPTG_A_S1b    18  571681  7.895900e-15  191   182     9   4.945055   \n",
       "495059  p3IPTG_A_S1b    18  569127  8.076710e-15  235   210    25  11.904762   \n",
       "443366   p3IPTG_A_S1    18  569125  0.000000e+00  534   463    71  15.334773   \n",
       "443009   p3IPTG_A_S1    18  566053  1.947940e-13  559   446   113  25.336323   \n",
       "\n",
       "                    REF              ALT                                  EFF  \\\n",
       "443045              CTA              TTT                          stop_gained   \n",
       "443082                A                C                          stop_gained   \n",
       "443002          TCCCTCG           TCCTCG                   frameshift_variant   \n",
       "443294                G                A                          stop_gained   \n",
       "443402                C                A                          stop_gained   \n",
       "...                 ...              ...                                  ...   \n",
       "443663    ATTTTTTTTGATA     ATTTTTTTGATA                   frameshift_variant   \n",
       "495218  ATTTTTTTTGATACT  ATTTCTTTAGAAACA                          stop_gained   \n",
       "495059           CAAACT            CAACT                   frameshift_variant   \n",
       "443366         ATCAAACT          GTCAACC  frameshift_variant&missense_variant   \n",
       "443009           CGCAAA           TGCTAT                          stop_gained   \n",
       "\n",
       "       IMPACT    AFF                MUT  ANN_TYPE  \n",
       "443045   HIGH  LacI4       p.LeuGly102*  cassette  \n",
       "443082   HIGH  LacI4           p.Tyr19*  cassette  \n",
       "443002   HIGH  LacI4         p.Gly230fs  cassette  \n",
       "443294   HIGH  cas9B         p.Gln1101*  cassette  \n",
       "443402   HIGH  cas9B          p.Glu790*  cassette  \n",
       "...       ...    ...                ...       ...  \n",
       "443663   HIGH  cas9B          p.Asn46fs  cassette  \n",
       "495218   HIGH  cas9B  p.SerIleLysLys42*  cassette  \n",
       "495059   HIGH  cas9B         p.Phe897fs  cassette  \n",
       "443366   HIGH  cas9B         p.Lys896fs  cassette  \n",
       "443009   HIGH  LacI4       p.LeuArg208*  cassette  \n",
       "\n",
       "[101 rows x 15 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example:\n",
    "## Notice the information is grouped by position, we will deal with this in section 3. \n",
    "data1[(data1['ANN_TYPE']=='cassette') & (data1['IMPACT']=='HIGH') & (data1['PASS']==18)].sort_values('FRAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>225393</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>786189</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>280</td>\n",
       "      <td>279</td>\n",
       "      <td>1</td>\n",
       "      <td>0.358423</td>\n",
       "      <td>TTAC</td>\n",
       "      <td>CTAT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas2 cat</td>\n",
       "      <td>p.Gln209*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>225340</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>785573</td>\n",
       "      <td>2.991160e-15</td>\n",
       "      <td>253</td>\n",
       "      <td>252</td>\n",
       "      <td>1</td>\n",
       "      <td>0.396825</td>\n",
       "      <td>GAAAAAAATCAC</td>\n",
       "      <td>GAAAAAAAATCAC</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas2 cat</td>\n",
       "      <td>p.Ile5fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>225346</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>785643</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>252</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>0.398406</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas2 cat</td>\n",
       "      <td>p.Gln26*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>264083</td>\n",
       "      <td>p3_B_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>566670</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>226</td>\n",
       "      <td>225</td>\n",
       "      <td>1</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>CTTTGC</td>\n",
       "      <td>CTTGC</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas1 LacI4</td>\n",
       "      <td>p.Lys3fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>225402</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>787394</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>150</td>\n",
       "      <td>149</td>\n",
       "      <td>1</td>\n",
       "      <td>0.671141</td>\n",
       "      <td>AGCG</td>\n",
       "      <td>TGCA</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas2 lacI4</td>\n",
       "      <td>p.TerArg373LeuGlnext*?</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>277049</td>\n",
       "      <td>p3_B_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>785709</td>\n",
       "      <td>1.702880e-15</td>\n",
       "      <td>258</td>\n",
       "      <td>256</td>\n",
       "      <td>2</td>\n",
       "      <td>0.781250</td>\n",
       "      <td>GTA</td>\n",
       "      <td>GA</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas2 cat</td>\n",
       "      <td>p.Val48fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>277050</td>\n",
       "      <td>p3_B_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>785714</td>\n",
       "      <td>1.642610e-15</td>\n",
       "      <td>255</td>\n",
       "      <td>253</td>\n",
       "      <td>2</td>\n",
       "      <td>0.790514</td>\n",
       "      <td>GAAAAATAAGC</td>\n",
       "      <td>GAAAATAAGC</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas2 cat</td>\n",
       "      <td>p.Asn51fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>209632</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>566676</td>\n",
       "      <td>1.994980e-14</td>\n",
       "      <td>226</td>\n",
       "      <td>224</td>\n",
       "      <td>2</td>\n",
       "      <td>0.892857</td>\n",
       "      <td>CATA</td>\n",
       "      <td>TATT</td>\n",
       "      <td>start_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas1 LacI4</td>\n",
       "      <td>p.Met1?</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>209630</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>566634</td>\n",
       "      <td>2.558530e-14</td>\n",
       "      <td>112</td>\n",
       "      <td>111</td>\n",
       "      <td>1</td>\n",
       "      <td>0.900901</td>\n",
       "      <td>AGGCT</td>\n",
       "      <td>AGGGCT</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas1 LacI4</td>\n",
       "      <td>p.Val16fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>225397</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>786318</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>98</td>\n",
       "      <td>97</td>\n",
       "      <td>1</td>\n",
       "      <td>1.030928</td>\n",
       "      <td>GCCTG</td>\n",
       "      <td>GCCCTG</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas2 lacI4</td>\n",
       "      <td>p.Val16fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>225356</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>785712</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>264</td>\n",
       "      <td>261</td>\n",
       "      <td>3</td>\n",
       "      <td>1.149425</td>\n",
       "      <td>AAGAAAA</td>\n",
       "      <td>TAGTAAT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas2 cat</td>\n",
       "      <td>p.LysLysAsn49*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN      FRAC  \\\n",
       "225393  p3_B_S1     3  786189  0.000000e+00  280   279     1  0.358423   \n",
       "225340  p3_B_S1     3  785573  2.991160e-15  253   252     1  0.396825   \n",
       "225346  p3_B_S1     3  785643  0.000000e+00  252   251     1  0.398406   \n",
       "264083  p3_B_S2     3  566670  0.000000e+00  226   225     1  0.444444   \n",
       "225402  p3_B_S1     3  787394  0.000000e+00  150   149     1  0.671141   \n",
       "277049  p3_B_S2     3  785709  1.702880e-15  258   256     2  0.781250   \n",
       "277050  p3_B_S2     3  785714  1.642610e-15  255   253     2  0.790514   \n",
       "209632  p3_B_S1     3  566676  1.994980e-14  226   224     2  0.892857   \n",
       "209630  p3_B_S1     3  566634  2.558530e-14  112   111     1  0.900901   \n",
       "225397  p3_B_S1     3  786318  0.000000e+00   98    97     1  1.030928   \n",
       "225356  p3_B_S1     3  785712  0.000000e+00  264   261     3  1.149425   \n",
       "\n",
       "                 REF            ALT                 EFF IMPACT         AFF  \\\n",
       "225393          TTAC           CTAT         stop_gained   HIGH    cas2 cat   \n",
       "225340  GAAAAAAATCAC  GAAAAAAAATCAC  frameshift_variant   HIGH    cas2 cat   \n",
       "225346             C              T         stop_gained   HIGH    cas2 cat   \n",
       "264083        CTTTGC          CTTGC  frameshift_variant   HIGH  cas1 LacI4   \n",
       "225402          AGCG           TGCA           stop_lost   HIGH  cas2 lacI4   \n",
       "277049           GTA             GA  frameshift_variant   HIGH    cas2 cat   \n",
       "277050   GAAAAATAAGC     GAAAATAAGC  frameshift_variant   HIGH    cas2 cat   \n",
       "209632          CATA           TATT          start_lost   HIGH  cas1 LacI4   \n",
       "209630         AGGCT         AGGGCT  frameshift_variant   HIGH  cas1 LacI4   \n",
       "225397         GCCTG         GCCCTG  frameshift_variant   HIGH  cas2 lacI4   \n",
       "225356       AAGAAAA        TAGTAAT         stop_gained   HIGH    cas2 cat   \n",
       "\n",
       "                           MUT  ANN_TYPE  \n",
       "225393               p.Gln209*  cassette  \n",
       "225340                p.Ile5fs  cassette  \n",
       "225346                p.Gln26*  cassette  \n",
       "264083                p.Lys3fs  cassette  \n",
       "225402  p.TerArg373LeuGlnext*?  cassette  \n",
       "277049               p.Val48fs  cassette  \n",
       "277050               p.Asn51fs  cassette  \n",
       "209632                 p.Met1?  cassette  \n",
       "209630               p.Val16fs  cassette  \n",
       "225397               p.Val16fs  cassette  \n",
       "225356          p.LysLysAsn49*  cassette  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For C20 we have snpcalls2 in the same format\n",
    "data2[(data2['ANN_TYPE']=='cassette') & (data2['IMPACT']=='HIGH') & (data2['PASS']==3)].sort_values('FRAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>124456</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>575262</td>\n",
       "      <td>4.520520e-15</td>\n",
       "      <td>613</td>\n",
       "      <td>612</td>\n",
       "      <td>1</td>\n",
       "      <td>0.163399</td>\n",
       "      <td>AGTA</td>\n",
       "      <td>TGTT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuLeu126*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124447</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>575215</td>\n",
       "      <td>3.737000e-15</td>\n",
       "      <td>612</td>\n",
       "      <td>611</td>\n",
       "      <td>1</td>\n",
       "      <td>0.163666</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Gln143*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124410</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>574999</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>587</td>\n",
       "      <td>586</td>\n",
       "      <td>1</td>\n",
       "      <td>0.170648</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Lys215*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124493</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>575509</td>\n",
       "      <td>7.700020e-15</td>\n",
       "      <td>577</td>\n",
       "      <td>576</td>\n",
       "      <td>1</td>\n",
       "      <td>0.173611</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Lys45*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124396</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>574915</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>534</td>\n",
       "      <td>533</td>\n",
       "      <td>1</td>\n",
       "      <td>0.187617</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Gln243*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124398</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>574924</td>\n",
       "      <td>1.536900e-15</td>\n",
       "      <td>530</td>\n",
       "      <td>529</td>\n",
       "      <td>1</td>\n",
       "      <td>0.189036</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Gln240*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46650</td>\n",
       "      <td>p2_8_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>575585</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>524</td>\n",
       "      <td>523</td>\n",
       "      <td>1</td>\n",
       "      <td>0.191205</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Tyr19*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46569</td>\n",
       "      <td>p2_8_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>574905</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>403</td>\n",
       "      <td>402</td>\n",
       "      <td>1</td>\n",
       "      <td>0.248756</td>\n",
       "      <td>TTTA</td>\n",
       "      <td>CTTT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuAsn245*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124342</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>574518</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>347</td>\n",
       "      <td>346</td>\n",
       "      <td>1</td>\n",
       "      <td>0.289017</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Ter375Trpext*?</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124343</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>574519</td>\n",
       "      <td>1.536900e-15</td>\n",
       "      <td>322</td>\n",
       "      <td>321</td>\n",
       "      <td>1</td>\n",
       "      <td>0.311526</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Ter375Gluext*?</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46530</td>\n",
       "      <td>p2_8_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>574557</td>\n",
       "      <td>1.659790e-16</td>\n",
       "      <td>315</td>\n",
       "      <td>314</td>\n",
       "      <td>1</td>\n",
       "      <td>0.318471</td>\n",
       "      <td>GCCAGTTG</td>\n",
       "      <td>ACCTGTTA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.GlnLeuAla360*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46526</td>\n",
       "      <td>p2_8_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>574524</td>\n",
       "      <td>5.695790e-15</td>\n",
       "      <td>278</td>\n",
       "      <td>277</td>\n",
       "      <td>1</td>\n",
       "      <td>0.361011</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Ter373Trpext*?</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46775</td>\n",
       "      <td>p2_8_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>580775</td>\n",
       "      <td>5.484850e-15</td>\n",
       "      <td>262</td>\n",
       "      <td>261</td>\n",
       "      <td>1</td>\n",
       "      <td>0.383142</td>\n",
       "      <td>TAT</td>\n",
       "      <td>AAA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.AspLys2*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124359</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>574657</td>\n",
       "      <td>4.897210e-15</td>\n",
       "      <td>473</td>\n",
       "      <td>471</td>\n",
       "      <td>2</td>\n",
       "      <td>0.424628</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Gln329*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46634</td>\n",
       "      <td>p2_8_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>575430</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>460</td>\n",
       "      <td>458</td>\n",
       "      <td>2</td>\n",
       "      <td>0.436681</td>\n",
       "      <td>TTGCCGGCCAATTGCTGAG</td>\n",
       "      <td>TTGTCGACCATTTACTGAG</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.GlnLeuAlaGly67*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46581</td>\n",
       "      <td>p2_8_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>574999</td>\n",
       "      <td>1.913590e-15</td>\n",
       "      <td>443</td>\n",
       "      <td>441</td>\n",
       "      <td>2</td>\n",
       "      <td>0.453515</td>\n",
       "      <td>TGTGTC</td>\n",
       "      <td>TCTGTT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.TrpHis213*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124481</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>575456</td>\n",
       "      <td>4.333470e-14</td>\n",
       "      <td>631</td>\n",
       "      <td>628</td>\n",
       "      <td>3</td>\n",
       "      <td>0.477707</td>\n",
       "      <td>GTTAG</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Asn62fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46655</td>\n",
       "      <td>p2_8_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>575628</td>\n",
       "      <td>2.622240e-15</td>\n",
       "      <td>403</td>\n",
       "      <td>401</td>\n",
       "      <td>2</td>\n",
       "      <td>0.498753</td>\n",
       "      <td>AATTC</td>\n",
       "      <td>TATTT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.GluLeu4*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124394</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>574905</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>541</td>\n",
       "      <td>538</td>\n",
       "      <td>3</td>\n",
       "      <td>0.557621</td>\n",
       "      <td>TTTA</td>\n",
       "      <td>CTTT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuAsn245*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124466</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>575349</td>\n",
       "      <td>1.441990e-14</td>\n",
       "      <td>692</td>\n",
       "      <td>688</td>\n",
       "      <td>4</td>\n",
       "      <td>0.581395</td>\n",
       "      <td>CTTGA</td>\n",
       "      <td>TTTTA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.SerArg97*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124387</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>574864</td>\n",
       "      <td>1.322980e-14</td>\n",
       "      <td>495</td>\n",
       "      <td>492</td>\n",
       "      <td>3</td>\n",
       "      <td>0.609756</td>\n",
       "      <td>GGTCATTCG</td>\n",
       "      <td>AGTCTTTCA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.AlaAsnAspGln257*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>124356</td>\n",
       "      <td>p2_8_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>574617</td>\n",
       "      <td>1.410350e-14</td>\n",
       "      <td>456</td>\n",
       "      <td>453</td>\n",
       "      <td>3</td>\n",
       "      <td>0.662252</td>\n",
       "      <td>AAGGTCGTTTTCCG</td>\n",
       "      <td>AAGGACGATTACCG</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LysThrThr339*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46525</td>\n",
       "      <td>p2_8_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>574519</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>263</td>\n",
       "      <td>261</td>\n",
       "      <td>2</td>\n",
       "      <td>0.766284</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Ter375Glnext*?</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46592</td>\n",
       "      <td>p2_8_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>575084</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>449</td>\n",
       "      <td>445</td>\n",
       "      <td>4</td>\n",
       "      <td>0.898876</td>\n",
       "      <td>TAAGT</td>\n",
       "      <td>TAGT</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Leu186fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46544</td>\n",
       "      <td>p2_8_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>574686</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>398</td>\n",
       "      <td>394</td>\n",
       "      <td>4</td>\n",
       "      <td>1.015228</td>\n",
       "      <td>AACTGCAATA</td>\n",
       "      <td>TACTTCTTTA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuGlnLeu317*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN      FRAC  \\\n",
       "124456  p2_8_S3     2  575262  4.520520e-15  613   612     1  0.163399   \n",
       "124447  p2_8_S3     2  575215  3.737000e-15  612   611     1  0.163666   \n",
       "124410  p2_8_S3     2  574999  0.000000e+00  587   586     1  0.170648   \n",
       "124493  p2_8_S3     2  575509  7.700020e-15  577   576     1  0.173611   \n",
       "124396  p2_8_S3     2  574915  0.000000e+00  534   533     1  0.187617   \n",
       "124398  p2_8_S3     2  574924  1.536900e-15  530   529     1  0.189036   \n",
       "46650   p2_8_S2     2  575585  0.000000e+00  524   523     1  0.191205   \n",
       "46569   p2_8_S2     2  574905  0.000000e+00  403   402     1  0.248756   \n",
       "124342  p2_8_S3     2  574518  0.000000e+00  347   346     1  0.289017   \n",
       "124343  p2_8_S3     2  574519  1.536900e-15  322   321     1  0.311526   \n",
       "46530   p2_8_S2     2  574557  1.659790e-16  315   314     1  0.318471   \n",
       "46526   p2_8_S2     2  574524  5.695790e-15  278   277     1  0.361011   \n",
       "46775   p2_8_S2     2  580775  5.484850e-15  262   261     1  0.383142   \n",
       "124359  p2_8_S3     2  574657  4.897210e-15  473   471     2  0.424628   \n",
       "46634   p2_8_S2     2  575430  0.000000e+00  460   458     2  0.436681   \n",
       "46581   p2_8_S2     2  574999  1.913590e-15  443   441     2  0.453515   \n",
       "124481  p2_8_S3     2  575456  4.333470e-14  631   628     3  0.477707   \n",
       "46655   p2_8_S2     2  575628  2.622240e-15  403   401     2  0.498753   \n",
       "124394  p2_8_S3     2  574905  0.000000e+00  541   538     3  0.557621   \n",
       "124466  p2_8_S3     2  575349  1.441990e-14  692   688     4  0.581395   \n",
       "124387  p2_8_S3     2  574864  1.322980e-14  495   492     3  0.609756   \n",
       "124356  p2_8_S3     2  574617  1.410350e-14  456   453     3  0.662252   \n",
       "46525   p2_8_S2     2  574519  0.000000e+00  263   261     2  0.766284   \n",
       "46592   p2_8_S2     2  575084  0.000000e+00  449   445     4  0.898876   \n",
       "46544   p2_8_S2     2  574686  0.000000e+00  398   394     4  1.015228   \n",
       "\n",
       "                        REF                  ALT                 EFF IMPACT  \\\n",
       "124456                 AGTA                 TGTT         stop_gained   HIGH   \n",
       "124447                    G                    A         stop_gained   HIGH   \n",
       "124410                    T                    A         stop_gained   HIGH   \n",
       "124493                    T                    A         stop_gained   HIGH   \n",
       "124396                    G                    A         stop_gained   HIGH   \n",
       "124398                    G                    A         stop_gained   HIGH   \n",
       "46650                     A                    T         stop_gained   HIGH   \n",
       "46569                  TTTA                 CTTT         stop_gained   HIGH   \n",
       "124342                    T                    C           stop_lost   HIGH   \n",
       "124343                    A                    C           stop_lost   HIGH   \n",
       "46530              GCCAGTTG             ACCTGTTA         stop_gained   HIGH   \n",
       "46526                     T                    C           stop_lost   HIGH   \n",
       "46775                   TAT                  AAA         stop_gained   HIGH   \n",
       "124359                    G                    A         stop_gained   HIGH   \n",
       "46634   TTGCCGGCCAATTGCTGAG  TTGTCGACCATTTACTGAG         stop_gained   HIGH   \n",
       "46581                TGTGTC               TCTGTT         stop_gained   HIGH   \n",
       "124481                GTTAG                 GTAG  frameshift_variant   HIGH   \n",
       "46655                 AATTC                TATTT         stop_gained   HIGH   \n",
       "124394                 TTTA                 CTTT         stop_gained   HIGH   \n",
       "124466                CTTGA                TTTTA         stop_gained   HIGH   \n",
       "124387            GGTCATTCG            AGTCTTTCA         stop_gained   HIGH   \n",
       "124356       AAGGTCGTTTTCCG       AAGGACGATTACCG         stop_gained   HIGH   \n",
       "46525                     A                    G           stop_lost   HIGH   \n",
       "46592                 TAAGT                 TAGT  frameshift_variant   HIGH   \n",
       "46544            AACTGCAATA           TACTTCTTTA         stop_gained   HIGH   \n",
       "\n",
       "          AFF                 MUT  ANN_TYPE  \n",
       "124456  LacI4        p.LeuLeu126*  cassette  \n",
       "124447  LacI4           p.Gln143*  cassette  \n",
       "124410  LacI4           p.Lys215*  cassette  \n",
       "124493  LacI4            p.Lys45*  cassette  \n",
       "124396  LacI4           p.Gln243*  cassette  \n",
       "124398  LacI4           p.Gln240*  cassette  \n",
       "46650   LacI4            p.Tyr19*  cassette  \n",
       "46569   LacI4        p.LeuAsn245*  cassette  \n",
       "124342  LacI4    p.Ter375Trpext*?  cassette  \n",
       "124343  LacI4    p.Ter375Gluext*?  cassette  \n",
       "46530   LacI4     p.GlnLeuAla360*  cassette  \n",
       "46526   LacI4    p.Ter373Trpext*?  cassette  \n",
       "46775   cas9B          p.AspLys2*  cassette  \n",
       "124359  LacI4           p.Gln329*  cassette  \n",
       "46634   LacI4   p.GlnLeuAlaGly67*  cassette  \n",
       "46581   LacI4        p.TrpHis213*  cassette  \n",
       "124481  LacI4           p.Asn62fs  cassette  \n",
       "46655   LacI4          p.GluLeu4*  cassette  \n",
       "124394  LacI4        p.LeuAsn245*  cassette  \n",
       "124466  LacI4         p.SerArg97*  cassette  \n",
       "124387  LacI4  p.AlaAsnAspGln257*  cassette  \n",
       "124356  LacI4     p.LysThrThr339*  cassette  \n",
       "46525   LacI4    p.Ter375Glnext*?  cassette  \n",
       "46592   LacI4          p.Leu186fs  cassette  \n",
       "46544   LacI4     p.LeuGlnLeu317*  cassette  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For C20 we have snpcalls2 in the same format\n",
    "data3[(data3['ANN_TYPE']=='cassette') & (data3['IMPACT']=='HIGH') & (data3['PASS']==2)].sort_values('FRAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save supplementary 1\n",
    "sup1 = './results/suptableS1.xlsx'\n",
    "if not os.path.isfile(sup1):\n",
    "    writer = pd.ExcelWriter(sup1, engine='xlsxwriter')\n",
    "    for c, data in zip(['C5_', 'C20_', 'C_26'], [data1, data2, data3]):\n",
    "        for sample in set(data['SAMPLE']):\n",
    "            subdata = data[data['SAMPLE']==sample].copy()\n",
    "            subdata.to_excel(writer, sheet_name=c+sample)\n",
    "    writer.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Rate and fraction of mutations in cassette compared to other distributions (supplementary figure 1)\n",
    "\n",
    "Explore the frequency at which we found a variant within the cassette compared to other types of annotations and also their representation within the population by means of the fraction values. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f6f954cbfe9e4254a1da1e8d9ae29eba",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def plot_percentage(df, genome=5):\n",
    "    \"\"\" Plot to show the percentage of variants mapping to each type of annotation \"\"\"\n",
    "    rs = {}\n",
    "    c = 0\n",
    "    lens = {}\n",
    "    lens['gene'] = len(ingene)\n",
    "    lens['intergenic'] = len(outgene)\n",
    "    lens['essential'] = len(posE)\n",
    "    lens['non-essential'] = len(posN)\n",
    "    if genome==5:\n",
    "        lens['chromosome'] = len(genome_C5)\n",
    "        lens['cassette'] = len(cas1)\n",
    "    elif genome==20:\n",
    "        lens['chromosome'] = len(genome_C20)\n",
    "        lens['cassette'] = len(cas12)\n",
    "    else:\n",
    "        lens['chromosome'] = len(genome_C26)\n",
    "        lens['cassette'] = len(cas33)\n",
    "    # Extract positions in passage 2\n",
    "    for pas in set(df['PASS']):\n",
    "        for sample in set(df[df['PASS']==pas]['SAMPLE']):\n",
    "            for impact in set(df['IMPACT']):\n",
    "                rs[c] = [pas, \n",
    "                         100*len(set(df[(df['PASS']==pas) & (df['IMPACT']==impact)  & (df['SAMPLE']==sample)]['POS']))/lens['chromosome'],\n",
    "                         sample, 'TOTAL', impact]\n",
    "                c+=1\n",
    "            rs[c] = [pas, 100*len(set(df[(df['PASS']==pas)  & (df['SAMPLE']==sample)]['POS']))/lens['chromosome'],sample, 'chromosome', 'TOTAL']\n",
    "            c+=1\n",
    "            for ann in set(df['ANN_TYPE']):\n",
    "                if ann in lens:\n",
    "                    rs[c] = [pas, \n",
    "                             100*len(set(df[(df['PASS']==pas) & (df['ANN_TYPE']==ann)  & (df['SAMPLE']==sample)]['POS']))/lens[ann],\n",
    "                             sample, ann, 'TOTAL']\n",
    "                    c+=1\n",
    "                    for impact in set(df['IMPACT']):\n",
    "                        rs[c] = [pas, \n",
    "                                 100*len(set(df[(df['PASS']==pas) & (df['ANN_TYPE']==ann) & (df['IMPACT']==impact)  & (df['SAMPLE']==sample)]['POS']))/lens[ann],\n",
    "                                 sample, ann, impact]\n",
    "                        c+=1\n",
    "    rs = pd.DataFrame.from_dict(rs, orient='index')\n",
    "    rs.columns = ['PASS', 'PERC', 'sample', 'annotation', 'IMPACT']\n",
    "    rs['HUE'] = rs['annotation']+rs['IMPACT']\n",
    "    return rs.sort_values('annotation')\n",
    "\n",
    "subdf1 = data1[(data1['FRAC']<100) & (data1['ALTN']>=2)].sort_values(['ANN_TYPE', 'PASS']).copy()\n",
    "subdf2 = data2[(data2['FRAC']<100) & (data2['ALTN']>=2)].sort_values(['ANN_TYPE', 'PASS']).copy()\n",
    "subdf3 = data3[(data3['FRAC']<100) & (data3['ALTN']>=2)].sort_values(['ANN_TYPE', 'PASS']).copy()\n",
    "\n",
    "plot1 = plot_percentage(subdf1)\n",
    "plot2 = plot_percentage(subdf2)\n",
    "plot3 = plot_percentage(subdf3)\n",
    "\n",
    "subdf1['log2Frac'] = np.log2(subdf1['FRAC'])\n",
    "subdf1['Condition'] = ['{}'.format(i) if i!=18 else '3IPTG' for i in subdf1['PASS']]\n",
    "subdf1['annotation'] = subdf1['ANN_TYPE']\n",
    "\n",
    "subdf2['log2Frac'] = np.log2(subdf2['FRAC'])\n",
    "subdf2['Condition'] = ['{}'.format(i) if i!=18 else '3IPTG' for i in subdf2['PASS']]\n",
    "subdf2['annotation'] = subdf2['ANN_TYPE']\n",
    "\n",
    "subdf3['log2Frac'] = np.log2(subdf3['FRAC'])\n",
    "subdf3['Condition'] = ['{}'.format(i) if i!=18 else '3IPTG' for i in subdf3['PASS']]\n",
    "subdf3['annotation'] = subdf3['ANN_TYPE']\n",
    "\n",
    "plt.close('all')\n",
    "plt.figure(figsize=(10, 10))\n",
    "#plt.subplot(3,3,1)\n",
    "#sns.countplot(x='Condition', hue='annotation', data=subdf1[(subdf1['ANN_TYPE'].isin(['cassette', 'gene', 'intergenic']))])\n",
    "\n",
    "plt.subplot(3,2,1)\n",
    "sns.barplot(x='PASS', y='PERC', hue='annotation', data=plot1[(plot1['IMPACT']=='TOTAL') & (plot1['annotation'].isin(['cassette', 'gene', 'intergenic']))])\n",
    "plt.xticks([0,1,2,3], ['2', '3', '15', '3IPTG'])\n",
    "plt.xlabel('Condition')\n",
    "plt.ylabel('Percentage')\n",
    "plt.title('Variant rate per base pair C5 samples')\n",
    "\n",
    "plt.subplot(3,2,2)\n",
    "sns.boxplot(y='log2Frac', x='Condition', hue='annotation', data=subdf1[(subdf1['log2Frac']>0) & (subdf1['ANN_TYPE'].isin(['cassette', 'gene', 'intergenic']))])\n",
    "plt.xlabel('Condition')\n",
    "plt.ylabel('log2(Fraction)')\n",
    "plt.title('Alternative Fraction C5 samples')\n",
    "\n",
    "##plt.ylim(0,3.2)\n",
    "#plt.subplot(3,3,4)\n",
    "#sns.countplot(x='PASS', hue='annotation', data=subdf2[(subdf2['ANN_TYPE'].isin(['cassette', 'gene', 'intergenic']))])\n",
    "#plt.xlabel('Condition')\n",
    "\n",
    "plt.subplot(3,2,3)\n",
    "sns.barplot(x='PASS', y='PERC', hue='annotation', data=plot2[(plot2['IMPACT']=='TOTAL') & (plot2['annotation'].isin(['cassette', 'gene', 'intergenic']))])\n",
    "plt.xlabel('Condition')\n",
    "plt.ylabel('Percentage')\n",
    "plt.title('Variant rate per base pair C20 samples')\n",
    "#plt.ylim(0,3.2)\n",
    "\n",
    "plt.subplot(3,2,4)\n",
    "sns.boxplot(y='log2Frac', x='PASS', hue='annotation', data=subdf2[(subdf2['log2Frac']>0) & (subdf2['ANN_TYPE'].isin(['cassette', 'gene', 'intergenic']))])\n",
    "plt.xlabel('Condition')\n",
    "plt.ylabel('log2(Fraction)')\n",
    "plt.title('Alternative Fraction  C20 samples')\n",
    "\n",
    "#plt.subplot(3,3,7)\n",
    "#sns.countplot(x='PASS', hue='annotation', data=subdf3[(subdf3['ANN_TYPE'].isin(['cassette', 'gene', 'intergenic']))])\n",
    "#plt.xlabel('Condition')\n",
    "\n",
    "plt.subplot(3,2,5)\n",
    "sns.barplot(x='PASS', y='PERC', hue='annotation', data=plot3[(plot3['IMPACT']=='TOTAL') & (plot3['annotation'].isin(['cassette', 'gene', 'intergenic']))])\n",
    "plt.xlabel('Condition')\n",
    "plt.ylabel('Percentage')\n",
    "plt.title('Variant rate per base pair C26 samples')\n",
    "#plt.ylim(0,3.2)\n",
    "\n",
    "plt.subplot(3,2,6)\n",
    "sns.boxplot(y='log2Frac', x='PASS', hue='annotation', data=subdf3[(subdf3['log2Frac']>0) & (subdf3['ANN_TYPE'].isin(['cassette', 'gene', 'intergenic']))])\n",
    "plt.xlabel('Condition')\n",
    "plt.ylabel('log2(Fraction)')\n",
    "plt.title('Alternative Fraction  C26 samples')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('./results/supfigS1.svg')\n",
    "plt.savefig('./results/supfigS1.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's calculate the statistics for this comparisons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import mannwhitneyu\n",
    "percentage_pvalues = {}\n",
    "fraction_pvalues = {}\n",
    "\n",
    "perc1 = plot1[(plot1['IMPACT']=='TOTAL') & (plot1['annotation'].isin(['cassette', 'gene', 'intergenic']))].copy()\n",
    "perc2 = plot2[(plot2['IMPACT']=='TOTAL') & (plot2['annotation'].isin(['cassette', 'gene', 'intergenic']))].copy()\n",
    "perc3 = plot3[(plot3['IMPACT']=='TOTAL') & (plot3['annotation'].isin(['cassette', 'gene', 'intergenic']))].copy()\n",
    "\n",
    "frac1 = subdf1[(subdf1['log2Frac']>0) & (subdf1['ANN_TYPE'].isin(['cassette', 'gene', 'intergenic']))].copy()\n",
    "frac2 = subdf2[(subdf2['log2Frac']>0) & (subdf2['ANN_TYPE'].isin(['cassette', 'gene', 'intergenic']))].copy()\n",
    "frac3 = subdf3[(subdf3['log2Frac']>0) & (subdf3['ANN_TYPE'].isin(['cassette', 'gene', 'intergenic']))].copy()\n",
    "\n",
    "c = 0\n",
    "for pas in set(perc1['PASS']):\n",
    "    x = perc1[(perc1['PASS']==pas) & (perc1['annotation']=='cassette')]['PERC']\n",
    "    for annot in ['gene', 'intergenic']:\n",
    "        y = perc1[(perc1['PASS']==pas) & (perc1['annotation']==annot)]['PERC']\n",
    "        percentage_pvalues[c] = ['C5', pas, annot, round(mannwhitneyu(x, y)[1]/2, 4)]\n",
    "        c+=1\n",
    "for pas in set(perc2['PASS']):\n",
    "    x = perc2[(perc2['PASS']==pas) & (perc2['annotation']=='cassette')]['PERC']\n",
    "    for annot in ['gene', 'intergenic']:\n",
    "        y = perc2[(perc2['PASS']==pas) & (perc2['annotation']==annot)]['PERC']\n",
    "        percentage_pvalues[c] = ['C20', pas, annot,round(mannwhitneyu(x, y)[1]/2, 4)]\n",
    "        c+=1\n",
    "for pas in set(perc3['PASS']):\n",
    "    x = perc3[(perc3['PASS']==pas) & (perc3['annotation']=='cassette')]['PERC']\n",
    "    for annot in ['gene', 'intergenic']:\n",
    "        y = perc3[(perc3['PASS']==pas) & (perc3['annotation']==annot)]['PERC']\n",
    "        percentage_pvalues[c] = ['C26', pas, annot,round(mannwhitneyu(x, y)[1]/2, 4)]\n",
    "        c+=1\n",
    "        \n",
    "c = 0\n",
    "for pas in set(frac1['PASS']):\n",
    "    x = frac1[(frac1['PASS']==pas) & (frac1['annotation']=='cassette')]['log2Frac']\n",
    "    for annot in ['gene', 'intergenic']:\n",
    "        y = frac1[(frac1['PASS']==pas) & (frac1['annotation']==annot)]['log2Frac']\n",
    "        fraction_pvalues[c] = ['C5', pas, annot, round(mannwhitneyu(x, y)[1]/2, 4)]\n",
    "        c+=1\n",
    "for pas in set(frac2['PASS']):\n",
    "    x = frac2[(frac2['PASS']==pas) & (frac2['annotation']=='cassette')]['log2Frac']\n",
    "    for annot in ['gene', 'intergenic']:\n",
    "        y = frac2[(frac2['PASS']==pas) & (frac2['annotation']==annot)]['log2Frac']\n",
    "        fraction_pvalues[c] = ['C20', pas, annot,round(mannwhitneyu(x, y)[1]/2, 4)]\n",
    "        c+=1\n",
    "for pas in set(frac3['PASS']):\n",
    "    x = frac3[(frac3['PASS']==pas) & (frac3['annotation']=='cassette')]['log2Frac']\n",
    "    for annot in ['gene', 'intergenic']:\n",
    "        y = frac3[(frac3['PASS']==pas) & (frac3['annotation']==annot)]['log2Frac']\n",
    "        fraction_pvalues[c] = ['C26', pas, annot,round(mannwhitneyu(x, y)[1]/2, 4)]\n",
    "        c+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average rate of mutations per base in the cassette 2.825228856974889\n",
      "Average rate of mutations per base in the intergenic 1.8428205685101453\n",
      "Average rate of mutations per base in the gene 1.4096396774343913\n",
      "Average rate of mutations per base in the cassette 0.9750566893424035\n",
      "Average rate of mutations per base in the intergenic 2.42446081044616\n",
      "Average rate of mutations per base in the gene 1.8454062368899455\n",
      "Average rate of mutations per base in the cassette 2.671957671957672\n",
      "Average rate of mutations per base in the intergenic 2.705737753470097\n",
      "Average rate of mutations per base in the gene 2.0816319395888687\n"
     ]
    }
   ],
   "source": [
    "print('Average rate of mutations per base in the cassette', np.mean(perc1[(perc1['annotation']=='cassette')]['PERC']))\n",
    "print('Average rate of mutations per base in the intergenic', np.mean(perc1[(perc1['annotation']=='intergenic')]['PERC']))\n",
    "print('Average rate of mutations per base in the gene', np.mean(perc1[(perc1['annotation']=='gene')]['PERC']))\n",
    "\n",
    "print('Average rate of mutations per base in the cassette', np.mean(perc2[(perc2['annotation']=='cassette')]['PERC']))\n",
    "print('Average rate of mutations per base in the intergenic', np.mean(perc2[(perc2['annotation']=='intergenic')]['PERC']))\n",
    "print('Average rate of mutations per base in the gene', np.mean(perc2[(perc2['annotation']=='gene')]['PERC']))\n",
    "\n",
    "print('Average rate of mutations per base in the cassette', np.mean(perc3[(perc3['annotation']=='cassette')]['PERC']))\n",
    "print('Average rate of mutations per base in the intergenic', np.mean(perc3[(perc3['annotation']=='intergenic')]['PERC']))\n",
    "print('Average rate of mutations per base in the gene', np.mean(perc3[(perc3['annotation']=='gene')]['PERC']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: ['C5', 18, 'gene', 0.0613],\n",
       " 1: ['C5', 18, 'intergenic', 0.1746],\n",
       " 2: ['C5', 2, 'gene', 0.0476],\n",
       " 3: ['C5', 2, 'intergenic', 0.0476],\n",
       " 4: ['C5', 3, 'gene', 0.1746],\n",
       " 5: ['C5', 3, 'intergenic', 0.1746],\n",
       " 6: ['C5', 15, 'gene', 0.0613],\n",
       " 7: ['C5', 15, 'intergenic', 0.0613],\n",
       " 8: ['C20', 2, 'gene', 0.0613],\n",
       " 9: ['C20', 2, 'intergenic', 0.0613],\n",
       " 10: ['C20', 3, 'gene', 0.0613],\n",
       " 11: ['C20', 3, 'intergenic', 0.0613],\n",
       " 12: ['C20', 15, 'gene', 0.0613],\n",
       " 13: ['C20', 15, 'intergenic', 0.0613],\n",
       " 14: ['C26', 2, 'gene', 0.0613],\n",
       " 15: ['C26', 2, 'intergenic', 0.1746],\n",
       " 16: ['C26', 15, 'gene', 0.1746],\n",
       " 17: ['C26', 15, 'intergenic', 0.1746]}"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "percentage_pvalues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average rate of mutations per base in the cassette 1.492537260055542\n",
      "Average rate of mutations per base in the intergenic 1.5564202070236206\n",
      "Average rate of mutations per base in the gene 1.470588207244873\n",
      "Average rate of mutations per base in the cassette 1.5151515\n",
      "Average rate of mutations per base in the intergenic 1.3513514\n",
      "Average rate of mutations per base in the gene 0.38646838\n",
      "Average rate of mutations per base in the cassette 1.3468167\n",
      "Average rate of mutations per base in the intergenic 1.3513514\n",
      "Average rate of mutations per base in the gene 0.36773175\n"
     ]
    }
   ],
   "source": [
    "print('Average rate of mutations per base in the cassette', np.percentile(frac1[(frac1['annotation']=='cassette')  & (frac1['log2Frac']>0)]['FRAC'], 50))\n",
    "print('Average rate of mutations per base in the intergenic', np.percentile(frac1[(frac1['annotation']=='intergenic') & (frac1['log2Frac']>0) ]['FRAC'], 50))\n",
    "print('Average rate of mutations per base in the gene', np.percentile(frac1[(frac1['annotation']=='gene')& (frac1['log2Frac']>0)]['FRAC'], 50))\n",
    "\n",
    "print('Average rate of mutations per base in the cassette', np.median(frac2[(frac2['annotation']=='cassette')]['FRAC']))\n",
    "print('Average rate of mutations per base in the intergenic', np.median(frac2[(frac2['annotation']=='intergenic')]['FRAC']))\n",
    "print('Average rate of mutations per base in the gene', np.median(frac2[(frac2['annotation']=='gene')]['log2Frac']))\n",
    "\n",
    "print('Average rate of mutations per base in the cassette', np.median(frac3[(frac3['annotation']=='cassette')]['FRAC']))\n",
    "print('Average rate of mutations per base in the intergenic', np.median(frac3[(frac3['annotation']=='intergenic')]['FRAC']))\n",
    "print('Average rate of mutations per base in the gene', np.median(frac3[(frac3['annotation']=='gene')]['log2Frac']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: ['C5', 18, 'gene', 0.0117],\n",
       " 1: ['C5', 18, 'intergenic', 0.2241],\n",
       " 2: ['C5', 2, 'gene', 0.0],\n",
       " 3: ['C5', 2, 'intergenic', 0.0],\n",
       " 4: ['C5', 3, 'gene', 0.1343],\n",
       " 5: ['C5', 3, 'intergenic', 0.0988],\n",
       " 6: ['C5', 15, 'gene', 0.1662],\n",
       " 7: ['C5', 15, 'intergenic', 0.0021],\n",
       " 8: ['C20', 2, 'gene', 0.0002],\n",
       " 9: ['C20', 2, 'intergenic', 0.0016],\n",
       " 10: ['C20', 3, 'gene', 0.0],\n",
       " 11: ['C20', 3, 'intergenic', 0.0],\n",
       " 12: ['C20', 15, 'gene', 0.0],\n",
       " 13: ['C20', 15, 'intergenic', 0.002],\n",
       " 14: ['C26', 2, 'gene', 0.2051],\n",
       " 15: ['C26', 2, 'intergenic', 0.1272],\n",
       " 16: ['C26', 15, 'gene', 0.0977],\n",
       " 17: ['C26', 15, 'intergenic', 0.03]}"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fraction_pvalues"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Location of the mutations within the cassettes (supplementary figure 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cfdc7fbad73347619b77fcc3fa54ac33",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "subdf1 = data1[(data1['FRAC']<100) & (data1['ALTN']>=2) & (data1['QUAL']>0.0)  & (data1['IMPACT']!='MODIFIER') & (~data1['AFF'].str.contains('pS'))].sort_values(['PASS', 'AFF']).copy()\n",
    "subdf2 = data2[(data2['FRAC']<100) & (data2['ALTN']>=2) & (data2['QUAL']>0.0) & (data2['IMPACT']!='MODIFIER')  & (~data2['AFF'].str.contains('pS'))].sort_values(['PASS', 'AFF']).copy() \n",
    "subdf3 = data3[(data3['FRAC']<100) & (data3['ALTN']>=2) & (data3['QUAL']>0.0) & (data3['IMPACT']!='MODIFIER')  & (~data3['AFF'].str.contains('pS'))].sort_values(['PASS', 'AFF']).copy() \n",
    "\n",
    "sorter = ['HIGH', 'MODERATE', 'LOW']\n",
    "subdf1.IMPACT = subdf1.IMPACT.astype(\"category\")\n",
    "subdf1.IMPACT.cat.set_categories(sorter, inplace=True)\n",
    "subdf2.IMPACT = subdf2.IMPACT.astype(\"category\")\n",
    "subdf2.IMPACT.cat.set_categories(sorter, inplace=True)\n",
    "subdf3.IMPACT = subdf3.IMPACT.astype(\"category\")\n",
    "subdf3.IMPACT.cat.set_categories(sorter, inplace=True)\n",
    "\n",
    "plot1 = plot_percentage(subdf1)\n",
    "plot2 = plot_percentage(subdf2)\n",
    "plot3 = plot_percentage(subdf3)\n",
    "\n",
    "subdf1['log2Frac'] = np.log2(subdf1['FRAC'])\n",
    "subdf1['Condition'] = ['{}'.format(i) if i!=18 else '3IPTG' for i in subdf1['PASS']]\n",
    "subdf1['annotation'] = subdf1['ANN_TYPE']\n",
    "\n",
    "subdf2['log2Frac'] = np.log2(subdf2['FRAC'])\n",
    "subdf2['Condition'] = ['{}'.format(i) if i!=18 else '3IPTG' for i in subdf2['PASS']]\n",
    "subdf2['annotation'] = subdf2['ANN_TYPE']\n",
    "\n",
    "subdf3['log2Frac'] = np.log2(subdf3['FRAC'])\n",
    "subdf3['Condition'] = ['{}'.format(i) if i!=18 else '3IPTG' for i in subdf3['PASS']]\n",
    "subdf3['annotation'] = subdf3['ANN_TYPE']\n",
    "\n",
    "plt.close('all')\n",
    "plt.figure(figsize=(10, 10))\n",
    "\n",
    "# C5\n",
    "plt.subplot(3,3,1)\n",
    "sns.countplot(x='AFF', hue='PASS', data=subdf1[(subdf1['ANN_TYPE'].isin(['cassette']))])\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.legend(title=\"Passage\")\n",
    "plt.xlabel('Element')\n",
    "\n",
    "plt.subplot(3,3,2)\n",
    "sns.swarmplot(y='FRAC', x='AFF', hue='PASS', data=subdf1[(subdf1['log2Frac']>0) & (subdf1['ANN_TYPE'].isin(['cassette']))])\n",
    "plt.xlabel('Element')\n",
    "plt.ylabel('Fraction')\n",
    "plt.legend(title=\"Passage\")\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.title('C5 samples')\n",
    "#plt.legend(bbox_to_anchor=(1.05, 1))\n",
    "\n",
    "plt.subplot(3,3,3)\n",
    "sns.swarmplot(x='AFF', y='FRAC', hue='IMPACT',data=subdf1[(subdf1['FRAC']>=1.5) & (subdf1['ANN_TYPE']=='cassette')  & (subdf1['IMPACT']!='MODIFIER')])\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.legend(title=\"Passage\")\n",
    "plt.xlabel('Element')\n",
    "plt.ylabel('Fraction')\n",
    "plt.legend(title=\"Impact\")\n",
    "\n",
    "#plt.ylim(0,3.2)\n",
    "plt.subplot(3,3,4)\n",
    "sns.countplot(x='AFF', hue='PASS', data=subdf2[(subdf2['ANN_TYPE'].isin(['cassette']))])\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.xlabel('Element')\n",
    "plt.legend(title=\"Passage\")\n",
    "\n",
    "plt.subplot(3,3,5)\n",
    "sns.swarmplot(y='FRAC', x='AFF', hue='PASS', data=subdf2[(subdf2['log2Frac']>0) & (subdf2['ANN_TYPE'].isin(['cassette']))])\n",
    "plt.xlabel('Element')\n",
    "plt.ylabel('Fraction')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.legend(title=\"Passage\")\n",
    "plt.title('C20 samples')\n",
    "#plt.legend(bbox_to_anchor=(1.05, 1))\n",
    "\n",
    "plt.subplot(3,3,6)\n",
    "sns.swarmplot(x='AFF', y='FRAC', hue='IMPACT',data=subdf2[(subdf2['FRAC']>=1.5) & (subdf2['ANN_TYPE']=='cassette')  & (subdf2['IMPACT']!='MODIFIER')])\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.xlabel('Element')\n",
    "plt.ylabel('Fraction')\n",
    "plt.legend(title=\"Impact\")\n",
    "\n",
    "# C26\n",
    "\n",
    "plt.subplot(3,3,7)\n",
    "sns.countplot(x='AFF', hue='PASS', data=subdf3[(subdf3['ANN_TYPE'].isin(['cassette']))])\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.xlabel('Element')\n",
    "plt.legend(title=\"Passage\")\n",
    "\n",
    "plt.subplot(3,3,8)\n",
    "sns.swarmplot(y='FRAC', x='AFF', hue='PASS', data=subdf3[(subdf3['log2Frac']>0) & (subdf3['ANN_TYPE'].isin(['cassette']))])\n",
    "plt.xlabel('Element')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.ylabel('Fraction')\n",
    "plt.title('C26 samples')\n",
    "plt.legend(title=\"Passage\")\n",
    "#plt.legend(bbox_to_anchor=(1.05, 1))\n",
    "\n",
    "plt.subplot(3,3, 9)\n",
    "sns.swarmplot(x='AFF', y='FRAC', hue='IMPACT',data=subdf3[(subdf3['FRAC']>=1.5) & (subdf3['ANN_TYPE']=='cassette')  & (subdf3['IMPACT']!='MODIFIER')])\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.xlabel('Element')\n",
    "plt.ylabel('Fraction')\n",
    "plt.legend(title=\"Impact\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('./results/supfigS2.svg')\n",
    "plt.savefig('./results/supfigS2.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>267963</td>\n",
       "      <td>p15_9_S7</td>\n",
       "      <td>15</td>\n",
       "      <td>580666</td>\n",
       "      <td>2.222470e-15</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>22.222221</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>cas2 cas9B</td>\n",
       "      <td>p.Asp39Gly</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN       FRAC REF  \\\n",
       "267963  p15_9_S7    15  580666  2.222470e-15   11     9     2  22.222221   T   \n",
       "\n",
       "       ALT               EFF    IMPACT         AFF         MUT  ANN_TYPE  \n",
       "267963   C  missense_variant  MODERATE  cas2 cas9B  p.Asp39Gly  cassette  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data3[(data3['PASS']==15) & (data3['FRAC']>19) & (data3['ANN_TYPE']=='cassette')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Exploration of the effect of the mutations explaining the escape rate (supplementary figure 3)\n",
    "\n",
    "Integrating the previous results we will explore the mutations that could mainly explain the rate of cell escaping the killswitch circuit (we use the median value to filter out non-significant positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "30dc9267316c47a68bc07ade0bcb221f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(10,5))\n",
    "plt.subplot(1,3,1)\n",
    "sns.swarmplot(x='AFF', y='FRAC', hue='IMPACT',data=subdf1[(subdf1['FRAC']>=1.5) & (subdf1['ANN_TYPE']=='cassette')  & (subdf1['IMPACT']!='MODIFIER')])\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.xlabel('Element')\n",
    "plt.ylabel('Fraction')\n",
    "\n",
    "plt.subplot(1,3,2)\n",
    "sns.swarmplot(x='AFF', y='FRAC', hue='IMPACT',data=subdf2[(subdf2['FRAC']>=1.5) & (subdf2['ANN_TYPE']=='cassette')  & (subdf2['IMPACT']!='MODIFIER')])\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.xlabel('Element')\n",
    "plt.ylabel('Fraction')\n",
    "\n",
    "plt.subplot(1,3,3)\n",
    "sns.swarmplot(x='AFF', y='FRAC', hue='IMPACT',data=subdf3[(subdf3['FRAC']>=1.5) & (subdf3['ANN_TYPE']=='cassette')  & (subdf3['IMPACT']!='MODIFIER')])\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.xlabel('Element')\n",
    "plt.ylabel('Fraction')\n",
    "plt.tight_layout()\n",
    "plt.savefig('./results/supfigS3.svg')\n",
    "plt.savefig('./results/supfigS3.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 677,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "      <th>TOP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>443045</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566372</td>\n",
       "      <td>2.184800e-15</td>\n",
       "      <td>677</td>\n",
       "      <td>676</td>\n",
       "      <td>1</td>\n",
       "      <td>0.147929</td>\n",
       "      <td>CTA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuGly102*</td>\n",
       "      <td>cassette</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443032</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566245</td>\n",
       "      <td>1.137610e-15</td>\n",
       "      <td>670</td>\n",
       "      <td>669</td>\n",
       "      <td>1</td>\n",
       "      <td>0.149477</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Ala145Asp</td>\n",
       "      <td>cassette</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443043</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566351</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>655</td>\n",
       "      <td>654</td>\n",
       "      <td>1</td>\n",
       "      <td>0.152905</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Met110Val</td>\n",
       "      <td>cassette</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443077</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566562</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>655</td>\n",
       "      <td>654</td>\n",
       "      <td>1</td>\n",
       "      <td>0.152905</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Ala39Ala</td>\n",
       "      <td>cassette</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443082</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566622</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>647</td>\n",
       "      <td>646</td>\n",
       "      <td>1</td>\n",
       "      <td>0.154799</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Tyr19*</td>\n",
       "      <td>cassette</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>494861</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>566416</td>\n",
       "      <td>3.471600e-14</td>\n",
       "      <td>281</td>\n",
       "      <td>255</td>\n",
       "      <td>26</td>\n",
       "      <td>10.196078</td>\n",
       "      <td>GGTGCGT</td>\n",
       "      <td>CGTTCTT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.HisAlaPro86GlnGluArg</td>\n",
       "      <td>cassette</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>495059</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>569127</td>\n",
       "      <td>8.076710e-15</td>\n",
       "      <td>235</td>\n",
       "      <td>210</td>\n",
       "      <td>25</td>\n",
       "      <td>11.904762</td>\n",
       "      <td>CAAACT</td>\n",
       "      <td>CAACT</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Phe897fs</td>\n",
       "      <td>cassette</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443366</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>569125</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>534</td>\n",
       "      <td>463</td>\n",
       "      <td>71</td>\n",
       "      <td>15.334773</td>\n",
       "      <td>ATCAAACT</td>\n",
       "      <td>GTCAACC</td>\n",
       "      <td>frameshift_variant&amp;missense_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Lys896fs</td>\n",
       "      <td>cassette</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>494831</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>566053</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>271</td>\n",
       "      <td>224</td>\n",
       "      <td>47</td>\n",
       "      <td>20.982143</td>\n",
       "      <td>CGCAA</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuArg208MetCys</td>\n",
       "      <td>cassette</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443009</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566053</td>\n",
       "      <td>1.947940e-13</td>\n",
       "      <td>559</td>\n",
       "      <td>446</td>\n",
       "      <td>113</td>\n",
       "      <td>25.336323</td>\n",
       "      <td>CGCAAA</td>\n",
       "      <td>TGCTAT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuArg208*</td>\n",
       "      <td>cassette</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1237 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN       FRAC  \\\n",
       "443045   p3IPTG_A_S1    18  566372  2.184800e-15  677   676     1   0.147929   \n",
       "443032   p3IPTG_A_S1    18  566245  1.137610e-15  670   669     1   0.149477   \n",
       "443043   p3IPTG_A_S1    18  566351  0.000000e+00  655   654     1   0.152905   \n",
       "443077   p3IPTG_A_S1    18  566562  0.000000e+00  655   654     1   0.152905   \n",
       "443082   p3IPTG_A_S1    18  566622  0.000000e+00  647   646     1   0.154799   \n",
       "...              ...   ...     ...           ...  ...   ...   ...        ...   \n",
       "494861  p3IPTG_A_S1b    18  566416  3.471600e-14  281   255    26  10.196078   \n",
       "495059  p3IPTG_A_S1b    18  569127  8.076710e-15  235   210    25  11.904762   \n",
       "443366   p3IPTG_A_S1    18  569125  0.000000e+00  534   463    71  15.334773   \n",
       "494831  p3IPTG_A_S1b    18  566053  0.000000e+00  271   224    47  20.982143   \n",
       "443009   p3IPTG_A_S1    18  566053  1.947940e-13  559   446   113  25.336323   \n",
       "\n",
       "             REF      ALT                                  EFF    IMPACT  \\\n",
       "443045       CTA      TTT                          stop_gained      HIGH   \n",
       "443032         G        T                     missense_variant  MODERATE   \n",
       "443043         T        C                     missense_variant  MODERATE   \n",
       "443077         T        G                   synonymous_variant       LOW   \n",
       "443082         A        C                          stop_gained      HIGH   \n",
       "...          ...      ...                                  ...       ...   \n",
       "494861   GGTGCGT  CGTTCTT                     missense_variant  MODERATE   \n",
       "495059    CAAACT    CAACT                   frameshift_variant      HIGH   \n",
       "443366  ATCAAACT  GTCAACC  frameshift_variant&missense_variant      HIGH   \n",
       "494831     CGCAA    CACAT                     missense_variant  MODERATE   \n",
       "443009    CGCAAA   TGCTAT                          stop_gained      HIGH   \n",
       "\n",
       "          AFF                     MUT  ANN_TYPE  TOP  \n",
       "443045  LacI4            p.LeuGly102*  cassette    0  \n",
       "443032  LacI4             p.Ala145Asp  cassette    0  \n",
       "443043  LacI4             p.Met110Val  cassette    0  \n",
       "443077  LacI4              p.Ala39Ala  cassette    0  \n",
       "443082  LacI4                p.Tyr19*  cassette    0  \n",
       "...       ...                     ...       ...  ...  \n",
       "494861  LacI4  p.HisAlaPro86GlnGluArg  cassette    1  \n",
       "495059  cas9B              p.Phe897fs  cassette    1  \n",
       "443366  cas9B              p.Lys896fs  cassette    1  \n",
       "494831  LacI4       p.LeuArg208MetCys  cassette    1  \n",
       "443009  LacI4            p.LeuArg208*  cassette    1  \n",
       "\n",
       "[1237 rows x 16 columns]"
      ]
     },
     "execution_count": 677,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected = data1[(data1['PASS']==18) & (data1['ANN_TYPE'].isin(['cassette']))].sort_values(['FRAC'])\n",
    "selected['TOP'] = [1 if z>=1.5 else 0 for z in selected['FRAC']]\n",
    "selected.to_excel('./results/suptableS2.xlsx')\n",
    "selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 678,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "      <th>TOP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>443045</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566372</td>\n",
       "      <td>2.184800e-15</td>\n",
       "      <td>677</td>\n",
       "      <td>676</td>\n",
       "      <td>1</td>\n",
       "      <td>0.147929</td>\n",
       "      <td>CTA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuGly102*</td>\n",
       "      <td>cassette</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443032</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566245</td>\n",
       "      <td>1.137610e-15</td>\n",
       "      <td>670</td>\n",
       "      <td>669</td>\n",
       "      <td>1</td>\n",
       "      <td>0.149477</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Ala145Asp</td>\n",
       "      <td>cassette</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443043</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566351</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>655</td>\n",
       "      <td>654</td>\n",
       "      <td>1</td>\n",
       "      <td>0.152905</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Met110Val</td>\n",
       "      <td>cassette</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443077</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566562</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>655</td>\n",
       "      <td>654</td>\n",
       "      <td>1</td>\n",
       "      <td>0.152905</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Ala39Ala</td>\n",
       "      <td>cassette</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443082</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566622</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>647</td>\n",
       "      <td>646</td>\n",
       "      <td>1</td>\n",
       "      <td>0.154799</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.Tyr19*</td>\n",
       "      <td>cassette</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>494861</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>566416</td>\n",
       "      <td>3.471600e-14</td>\n",
       "      <td>281</td>\n",
       "      <td>255</td>\n",
       "      <td>26</td>\n",
       "      <td>10.196078</td>\n",
       "      <td>GGTGCGT</td>\n",
       "      <td>CGTTCTT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.HisAlaPro86GlnGluArg</td>\n",
       "      <td>cassette</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>495059</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>569127</td>\n",
       "      <td>8.076710e-15</td>\n",
       "      <td>235</td>\n",
       "      <td>210</td>\n",
       "      <td>25</td>\n",
       "      <td>11.904762</td>\n",
       "      <td>CAAACT</td>\n",
       "      <td>CAACT</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Phe897fs</td>\n",
       "      <td>cassette</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443366</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>569125</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>534</td>\n",
       "      <td>463</td>\n",
       "      <td>71</td>\n",
       "      <td>15.334773</td>\n",
       "      <td>ATCAAACT</td>\n",
       "      <td>GTCAACC</td>\n",
       "      <td>frameshift_variant&amp;missense_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Lys896fs</td>\n",
       "      <td>cassette</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>494831</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>566053</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>271</td>\n",
       "      <td>224</td>\n",
       "      <td>47</td>\n",
       "      <td>20.982143</td>\n",
       "      <td>CGCAA</td>\n",
       "      <td>CACAT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuArg208MetCys</td>\n",
       "      <td>cassette</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443009</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566053</td>\n",
       "      <td>1.947940e-13</td>\n",
       "      <td>559</td>\n",
       "      <td>446</td>\n",
       "      <td>113</td>\n",
       "      <td>25.336323</td>\n",
       "      <td>CGCAAA</td>\n",
       "      <td>TGCTAT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuArg208*</td>\n",
       "      <td>cassette</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1237 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN       FRAC  \\\n",
       "443045   p3IPTG_A_S1    18  566372  2.184800e-15  677   676     1   0.147929   \n",
       "443032   p3IPTG_A_S1    18  566245  1.137610e-15  670   669     1   0.149477   \n",
       "443043   p3IPTG_A_S1    18  566351  0.000000e+00  655   654     1   0.152905   \n",
       "443077   p3IPTG_A_S1    18  566562  0.000000e+00  655   654     1   0.152905   \n",
       "443082   p3IPTG_A_S1    18  566622  0.000000e+00  647   646     1   0.154799   \n",
       "...              ...   ...     ...           ...  ...   ...   ...        ...   \n",
       "494861  p3IPTG_A_S1b    18  566416  3.471600e-14  281   255    26  10.196078   \n",
       "495059  p3IPTG_A_S1b    18  569127  8.076710e-15  235   210    25  11.904762   \n",
       "443366   p3IPTG_A_S1    18  569125  0.000000e+00  534   463    71  15.334773   \n",
       "494831  p3IPTG_A_S1b    18  566053  0.000000e+00  271   224    47  20.982143   \n",
       "443009   p3IPTG_A_S1    18  566053  1.947940e-13  559   446   113  25.336323   \n",
       "\n",
       "             REF      ALT                                  EFF    IMPACT  \\\n",
       "443045       CTA      TTT                          stop_gained      HIGH   \n",
       "443032         G        T                     missense_variant  MODERATE   \n",
       "443043         T        C                     missense_variant  MODERATE   \n",
       "443077         T        G                   synonymous_variant       LOW   \n",
       "443082         A        C                          stop_gained      HIGH   \n",
       "...          ...      ...                                  ...       ...   \n",
       "494861   GGTGCGT  CGTTCTT                     missense_variant  MODERATE   \n",
       "495059    CAAACT    CAACT                   frameshift_variant      HIGH   \n",
       "443366  ATCAAACT  GTCAACC  frameshift_variant&missense_variant      HIGH   \n",
       "494831     CGCAA    CACAT                     missense_variant  MODERATE   \n",
       "443009    CGCAAA   TGCTAT                          stop_gained      HIGH   \n",
       "\n",
       "          AFF                     MUT  ANN_TYPE  TOP  \n",
       "443045  LacI4            p.LeuGly102*  cassette    0  \n",
       "443032  LacI4             p.Ala145Asp  cassette    0  \n",
       "443043  LacI4             p.Met110Val  cassette    0  \n",
       "443077  LacI4              p.Ala39Ala  cassette    0  \n",
       "443082  LacI4                p.Tyr19*  cassette    0  \n",
       "...       ...                     ...       ...  ...  \n",
       "494861  LacI4  p.HisAlaPro86GlnGluArg  cassette    1  \n",
       "495059  cas9B              p.Phe897fs  cassette    1  \n",
       "443366  cas9B              p.Lys896fs  cassette    1  \n",
       "494831  LacI4       p.LeuArg208MetCys  cassette    1  \n",
       "443009  LacI4            p.LeuArg208*  cassette    1  \n",
       "\n",
       "[1237 rows x 16 columns]"
      ]
     },
     "execution_count": 678,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected[selected['POS'].isin([566423, 571681, 566053])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 681,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>54232</td>\n",
       "      <td>p2_6_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>566587</td>\n",
       "      <td>3.889600e-06</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>33.333332</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>cas1 LacI4</td>\n",
       "      <td>p.Thr31Ile</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>163230</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>780968</td>\n",
       "      <td>4.857580e+03</td>\n",
       "      <td>156</td>\n",
       "      <td>1</td>\n",
       "      <td>155</td>\n",
       "      <td>15500.000000</td>\n",
       "      <td>AGTCC</td>\n",
       "      <td>CGTCC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>cas2 regulator gRNA5</td>\n",
       "      <td>c.-4520T&gt;G</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>163253</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>781179</td>\n",
       "      <td>6.057650e-15</td>\n",
       "      <td>80</td>\n",
       "      <td>71</td>\n",
       "      <td>9</td>\n",
       "      <td>12.676056</td>\n",
       "      <td>GACCCAACTGCCACGAAGTTTT</td>\n",
       "      <td>GACCCAACTGCCTTGATGTTAT</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>cas2 regulator gRNA2</td>\n",
       "      <td>c.-4751_-4743delAAACTTCGTinsTAACATCAA</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>163342</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>786151</td>\n",
       "      <td>5.734500e+03</td>\n",
       "      <td>182</td>\n",
       "      <td>1</td>\n",
       "      <td>181</td>\n",
       "      <td>18100.000000</td>\n",
       "      <td>TCT</td>\n",
       "      <td>TTT</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>cas2 cat</td>\n",
       "      <td>p.Val195Val</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>163362</td>\n",
       "      <td>p2_6_S4</td>\n",
       "      <td>2</td>\n",
       "      <td>786347</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>14</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>27.272728</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>cas2 lacI4</td>\n",
       "      <td>p.Tyr24Cys</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>225311</td>\n",
       "      <td>p3_B_S1</td>\n",
       "      <td>3</td>\n",
       "      <td>780968</td>\n",
       "      <td>6.689500e+03</td>\n",
       "      <td>214</td>\n",
       "      <td>2</td>\n",
       "      <td>212</td>\n",
       "      <td>10600.000000</td>\n",
       "      <td>AGTCC</td>\n",
       "      <td>CGTCC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>cas2 regulator gRNA5</td>\n",
       "      <td>c.-4520T&gt;G</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>277012</td>\n",
       "      <td>p3_B_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>780964</td>\n",
       "      <td>5.288950e+03</td>\n",
       "      <td>170</td>\n",
       "      <td>2</td>\n",
       "      <td>168</td>\n",
       "      <td>8400.000000</td>\n",
       "      <td>GGCTAGTCC</td>\n",
       "      <td>GGCTCGTCC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>cas2 regulator gRNA5</td>\n",
       "      <td>c.-4520T&gt;G</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>277062</td>\n",
       "      <td>p3_B_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>786151</td>\n",
       "      <td>7.229180e+03</td>\n",
       "      <td>228</td>\n",
       "      <td>1</td>\n",
       "      <td>227</td>\n",
       "      <td>22700.000000</td>\n",
       "      <td>TCTGTG</td>\n",
       "      <td>TTTGTG</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>cas2 cat</td>\n",
       "      <td>p.Val195Val</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>413818</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>780968</td>\n",
       "      <td>3.700950e+03</td>\n",
       "      <td>118</td>\n",
       "      <td>1</td>\n",
       "      <td>117</td>\n",
       "      <td>11700.000000</td>\n",
       "      <td>AGTCC</td>\n",
       "      <td>CGTCC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>cas2 regulator gRNA5</td>\n",
       "      <td>c.-4520T&gt;G</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>413844</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>785418</td>\n",
       "      <td>4.709100e-15</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666666</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>cas2 cas9B</td>\n",
       "      <td>p.Thr1339Thr</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>413909</td>\n",
       "      <td>p15_7_S5</td>\n",
       "      <td>15</td>\n",
       "      <td>786347</td>\n",
       "      <td>6.060810e-12</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>66.666664</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>cas2 lacI4</td>\n",
       "      <td>p.Tyr24Cys</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN          FRAC  \\\n",
       "54232    p2_6_S2     2  566587  3.889600e-06    4     3     1     33.333332   \n",
       "163230   p2_6_S4     2  780968  4.857580e+03  156     1   155  15500.000000   \n",
       "163253   p2_6_S4     2  781179  6.057650e-15   80    71     9     12.676056   \n",
       "163342   p2_6_S4     2  786151  5.734500e+03  182     1   181  18100.000000   \n",
       "163362   p2_6_S4     2  786347  0.000000e+00   14    11     3     27.272728   \n",
       "225311   p3_B_S1     3  780968  6.689500e+03  214     2   212  10600.000000   \n",
       "277012   p3_B_S2     3  780964  5.288950e+03  170     2   168   8400.000000   \n",
       "277062   p3_B_S2     3  786151  7.229180e+03  228     1   227  22700.000000   \n",
       "413818  p15_7_S5    15  780968  3.700950e+03  118     1   117  11700.000000   \n",
       "413844  p15_7_S5    15  785418  4.709100e-15    7     6     1     16.666666   \n",
       "413909  p15_7_S5    15  786347  6.060810e-12   10     6     4     66.666664   \n",
       "\n",
       "                           REF                     ALT                    EFF  \\\n",
       "54232                        G                       A       missense_variant   \n",
       "163230                   AGTCC                   CGTCC  upstream_gene_variant   \n",
       "163253  GACCCAACTGCCACGAAGTTTT  GACCCAACTGCCTTGATGTTAT  upstream_gene_variant   \n",
       "163342                     TCT                     TTT     synonymous_variant   \n",
       "163362                       A                       G       missense_variant   \n",
       "225311                   AGTCC                   CGTCC  upstream_gene_variant   \n",
       "277012               GGCTAGTCC               GGCTCGTCC  upstream_gene_variant   \n",
       "277062                  TCTGTG                  TTTGTG     synonymous_variant   \n",
       "413818                   AGTCC                   CGTCC  upstream_gene_variant   \n",
       "413844                       A                       C     synonymous_variant   \n",
       "413909                       A                       G       missense_variant   \n",
       "\n",
       "          IMPACT                   AFF                                    MUT  \\\n",
       "54232   MODERATE            cas1 LacI4                             p.Thr31Ile   \n",
       "163230  MODIFIER  cas2 regulator gRNA5                             c.-4520T>G   \n",
       "163253  MODIFIER  cas2 regulator gRNA2  c.-4751_-4743delAAACTTCGTinsTAACATCAA   \n",
       "163342       LOW              cas2 cat                            p.Val195Val   \n",
       "163362  MODERATE            cas2 lacI4                             p.Tyr24Cys   \n",
       "225311  MODIFIER  cas2 regulator gRNA5                             c.-4520T>G   \n",
       "277012  MODIFIER  cas2 regulator gRNA5                             c.-4520T>G   \n",
       "277062       LOW              cas2 cat                            p.Val195Val   \n",
       "413818  MODIFIER  cas2 regulator gRNA5                             c.-4520T>G   \n",
       "413844       LOW            cas2 cas9B                           p.Thr1339Thr   \n",
       "413909  MODERATE            cas2 lacI4                             p.Tyr24Cys   \n",
       "\n",
       "        ANN_TYPE  \n",
       "54232   cassette  \n",
       "163230  cassette  \n",
       "163253  cassette  \n",
       "163342  cassette  \n",
       "163362  cassette  \n",
       "225311  cassette  \n",
       "277012  cassette  \n",
       "277062  cassette  \n",
       "413818  cassette  \n",
       "413844  cassette  \n",
       "413909  cassette  "
      ]
     },
     "execution_count": 681,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data2[(data2['ANN_TYPE']=='cassette') & (data2['FRAC']>10)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 661,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>494840</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>566136</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>243</td>\n",
       "      <td>238</td>\n",
       "      <td>5</td>\n",
       "      <td>2.100840</td>\n",
       "      <td>TAAGCGGGTCCCATCTTCGT</td>\n",
       "      <td>TTAGCCGGTCCCATCTTCGT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.ArgLeu180*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>494818</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>565929</td>\n",
       "      <td>6.208090e-15</td>\n",
       "      <td>232</td>\n",
       "      <td>227</td>\n",
       "      <td>5</td>\n",
       "      <td>2.202643</td>\n",
       "      <td>AACAATCCCCTCATTTAA</td>\n",
       "      <td>AACAATCCCCTTTTTTTA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuAsnGlu245*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443663</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>571681</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>481</td>\n",
       "      <td>469</td>\n",
       "      <td>12</td>\n",
       "      <td>2.558635</td>\n",
       "      <td>ATTTTTTTTGATA</td>\n",
       "      <td>ATTTTTTTGATA</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Asn46fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>495218</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>571681</td>\n",
       "      <td>7.895900e-15</td>\n",
       "      <td>191</td>\n",
       "      <td>182</td>\n",
       "      <td>9</td>\n",
       "      <td>4.945055</td>\n",
       "      <td>ATTTTTTTTGATACT</td>\n",
       "      <td>ATTTCTTTAGAAACA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.SerIleLysLys42*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>495059</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>569127</td>\n",
       "      <td>8.076710e-15</td>\n",
       "      <td>235</td>\n",
       "      <td>210</td>\n",
       "      <td>25</td>\n",
       "      <td>11.904762</td>\n",
       "      <td>CAAACT</td>\n",
       "      <td>CAACT</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Phe897fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443366</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>569125</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>534</td>\n",
       "      <td>463</td>\n",
       "      <td>71</td>\n",
       "      <td>15.334773</td>\n",
       "      <td>ATCAAACT</td>\n",
       "      <td>GTCAACC</td>\n",
       "      <td>frameshift_variant&amp;missense_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.Lys896fs</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>443009</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>566053</td>\n",
       "      <td>1.947940e-13</td>\n",
       "      <td>559</td>\n",
       "      <td>446</td>\n",
       "      <td>113</td>\n",
       "      <td>25.336323</td>\n",
       "      <td>CGCAAA</td>\n",
       "      <td>TGCTAT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LacI4</td>\n",
       "      <td>p.LeuArg208*</td>\n",
       "      <td>cassette</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN       FRAC  \\\n",
       "494840  p3IPTG_A_S1b    18  566136  0.000000e+00  243   238     5   2.100840   \n",
       "494818  p3IPTG_A_S1b    18  565929  6.208090e-15  232   227     5   2.202643   \n",
       "443663   p3IPTG_A_S1    18  571681  0.000000e+00  481   469    12   2.558635   \n",
       "495218  p3IPTG_A_S1b    18  571681  7.895900e-15  191   182     9   4.945055   \n",
       "495059  p3IPTG_A_S1b    18  569127  8.076710e-15  235   210    25  11.904762   \n",
       "443366   p3IPTG_A_S1    18  569125  0.000000e+00  534   463    71  15.334773   \n",
       "443009   p3IPTG_A_S1    18  566053  1.947940e-13  559   446   113  25.336323   \n",
       "\n",
       "                         REF                   ALT  \\\n",
       "494840  TAAGCGGGTCCCATCTTCGT  TTAGCCGGTCCCATCTTCGT   \n",
       "494818    AACAATCCCCTCATTTAA    AACAATCCCCTTTTTTTA   \n",
       "443663         ATTTTTTTTGATA          ATTTTTTTGATA   \n",
       "495218       ATTTTTTTTGATACT       ATTTCTTTAGAAACA   \n",
       "495059                CAAACT                 CAACT   \n",
       "443366              ATCAAACT               GTCAACC   \n",
       "443009                CGCAAA                TGCTAT   \n",
       "\n",
       "                                        EFF IMPACT    AFF                MUT  \\\n",
       "494840                          stop_gained   HIGH  LacI4       p.ArgLeu180*   \n",
       "494818                          stop_gained   HIGH  LacI4    p.LeuAsnGlu245*   \n",
       "443663                   frameshift_variant   HIGH  cas9B          p.Asn46fs   \n",
       "495218                          stop_gained   HIGH  cas9B  p.SerIleLysLys42*   \n",
       "495059                   frameshift_variant   HIGH  cas9B         p.Phe897fs   \n",
       "443366  frameshift_variant&missense_variant   HIGH  cas9B         p.Lys896fs   \n",
       "443009                          stop_gained   HIGH  LacI4       p.LeuArg208*   \n",
       "\n",
       "        ANN_TYPE  \n",
       "494840  cassette  \n",
       "494818  cassette  \n",
       "443663  cassette  \n",
       "495218  cassette  \n",
       "495059  cassette  \n",
       "443366  cassette  \n",
       "443009  cassette  "
      ]
     },
     "execution_count": 661,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected[selected['IMPACT']=='HIGH']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TOPLOT "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- C5 fractions versus percentage of variants per annotation type\n",
    "- C5 cassette separating LacI4, Cas9B, gRNA/s only and impact+fraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "19a818e8045949c297190c22b0d7f17e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure()\n",
    "c = 1\n",
    "for pas in [2, 3, 15, 18]:\n",
    "    plt.subplot(4,1, c)\n",
    "    sns.scatterplot(x='POS', y='FRAC', hue='IMPACT', data=data1[(data1['FRAC']>2.5) & (data1['PASS']==pas) & (~data1['ANN_TYPE'].isin(['essential', 'non-essential']))])\n",
    "    c+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7613b6048aa94fb6a5a7a1f05db306ff",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure()\n",
    "c = 1\n",
    "for pas in [2, 3, 15]:\n",
    "    plt.subplot(4,1, c)\n",
    "    sns.scatterplot(x='POS', y='FRAC', hue='IMPACT', data=data2[(data2['FRAC']>2.5) & (data2['FRAC']<100) & (data2['PASS']==pas) & (~data2['ANN_TYPE'].isin(['essential', 'non-essential']))])\n",
    "    c+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NOT USED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 635,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "      <th>log2Frac</th>\n",
       "      <th>Condition</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>20557</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>629173</td>\n",
       "      <td>1.383230e-14</td>\n",
       "      <td>30</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>76.470589</td>\n",
       "      <td>GTTTTTTTTTTTTTTTTAGTTTGAAC</td>\n",
       "      <td>GTTTTTTTTTATTTATAAGTTTGAAC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN508</td>\n",
       "      <td>c.-2605_-2599delAAAAAAAinsTATAAAT</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>6.256833</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>33962</td>\n",
       "      <td>p2_2_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>113969</td>\n",
       "      <td>2.790510e-07</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN094</td>\n",
       "      <td>c.-2292T&gt;G</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>5.643856</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>35494</td>\n",
       "      <td>p2_2_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>141270</td>\n",
       "      <td>4.924100e-14</td>\n",
       "      <td>59</td>\n",
       "      <td>40</td>\n",
       "      <td>19</td>\n",
       "      <td>47.500000</td>\n",
       "      <td>CAGAGAGAGAGAGAGAGAGAGC</td>\n",
       "      <td>CAGAGAGAGAGAGAGAGAGC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN109</td>\n",
       "      <td>c.-624_-623delAG</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>5.569856</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>57248</td>\n",
       "      <td>p2_2_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>528761</td>\n",
       "      <td>2.688110e-14</td>\n",
       "      <td>45</td>\n",
       "      <td>33</td>\n",
       "      <td>12</td>\n",
       "      <td>36.363636</td>\n",
       "      <td>GTTTTTTTTTTTTTTTTTGAAAGA</td>\n",
       "      <td>GTTTTTTTTTTTTTTTTGAAAGA</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN435</td>\n",
       "      <td>c.-3956delA</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>5.184424</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>63072</td>\n",
       "      <td>p2_2_S2</td>\n",
       "      <td>2</td>\n",
       "      <td>629170</td>\n",
       "      <td>2.953720e-15</td>\n",
       "      <td>57</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>CTAGTTTTTTTTTTTTTTTTAGTTTGAAC</td>\n",
       "      <td>TTAGATTTTTTTTTTTTTTTTAGTTTGAAC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN508</td>\n",
       "      <td>c.-2614_-2586delGTTCAAACTAAAAAAAAAAAAAAAACTAGi...</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>5.643856</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>85859</td>\n",
       "      <td>p2_2_S3</td>\n",
       "      <td>2</td>\n",
       "      <td>141267</td>\n",
       "      <td>1.219870e-13</td>\n",
       "      <td>159</td>\n",
       "      <td>97</td>\n",
       "      <td>62</td>\n",
       "      <td>63.917526</td>\n",
       "      <td>TCTCAGAGAGAGAGAGAGAGAGAGC</td>\n",
       "      <td>ACTCAGAGAGAGAGAGAGAGAGAGAGC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN109</td>\n",
       "      <td>c.-646_-622delTCTCAGAGAGAGAGAGAGAGAGAGCinsACTC...</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>5.998140</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>161975</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>195424</td>\n",
       "      <td>4.551630e-13</td>\n",
       "      <td>413</td>\n",
       "      <td>239</td>\n",
       "      <td>174</td>\n",
       "      <td>72.803345</td>\n",
       "      <td>TTTCCAAAAAAAAAAAAAAAGTAAAATAGAAAAGC</td>\n",
       "      <td>ATTCTGAAAAAAAAAAAAAAAAGTAAAATAGAAAAGC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN148</td>\n",
       "      <td>c.-97_-63delTTTCCAAAAAAAAAAAAAAAGTAAAATAGAAAAG...</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>6.185933</td>\n",
       "      <td>3</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>214268</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>629169</td>\n",
       "      <td>2.384300e-13</td>\n",
       "      <td>355</td>\n",
       "      <td>212</td>\n",
       "      <td>143</td>\n",
       "      <td>67.452827</td>\n",
       "      <td>TCTAGTTTTTTTTTTTTTTTTAGTTTGAAC</td>\n",
       "      <td>TCTAGTTTTTTTTTTTTCTATAGTTTGAAC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN508</td>\n",
       "      <td>c.-2604_-2602delAAAinsTAG</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>6.075807</td>\n",
       "      <td>3</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>274642</td>\n",
       "      <td>p3_D_S3</td>\n",
       "      <td>3</td>\n",
       "      <td>528755</td>\n",
       "      <td>8.763400e-14</td>\n",
       "      <td>140</td>\n",
       "      <td>84</td>\n",
       "      <td>56</td>\n",
       "      <td>66.666664</td>\n",
       "      <td>TCAAACGTTTTTTTTTTTTTTTTTGAAAGAAATTGATTGCT</td>\n",
       "      <td>CCAATCGTTTTTTTTTTTTTTTTTGAAAGAAATTGATTGCT</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN435</td>\n",
       "      <td>c.-3937_-3933delTTTGAinsATTGG</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>6.058894</td>\n",
       "      <td>3</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>280252</td>\n",
       "      <td>p3_D_S3</td>\n",
       "      <td>3</td>\n",
       "      <td>618608</td>\n",
       "      <td>8.492520e-09</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN504</td>\n",
       "      <td>c.-702A&gt;G</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>5.321928</td>\n",
       "      <td>3</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>406624</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>141270</td>\n",
       "      <td>2.114290e-13</td>\n",
       "      <td>235</td>\n",
       "      <td>158</td>\n",
       "      <td>77</td>\n",
       "      <td>48.734177</td>\n",
       "      <td>CAGAGAGAGAGAGAGAGAGAGC</td>\n",
       "      <td>CAGAGAGAAAAAGAGAGAGAGC</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN109</td>\n",
       "      <td>c.-635_-633delGAGinsAAA</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>5.606862</td>\n",
       "      <td>3IPTG</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>447613</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>618608</td>\n",
       "      <td>2.102690e-14</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>28.571428</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN504</td>\n",
       "      <td>c.-702A&gt;G</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>4.836501</td>\n",
       "      <td>3IPTG</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>448420</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>629167</td>\n",
       "      <td>7.256610e-14</td>\n",
       "      <td>246</td>\n",
       "      <td>139</td>\n",
       "      <td>107</td>\n",
       "      <td>76.978416</td>\n",
       "      <td>ATTCTAGTTTTTTTTTTTTTTTTAGTTTGAACTCAAAA</td>\n",
       "      <td>TTTTTAATTTTTTTTTTTTTTAGTTTGAACTCAAAA</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN508</td>\n",
       "      <td>c.-2620_-2583delTTTTGAGTTCAAACTAAAAAAAAAAAAAAA...</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>6.266382</td>\n",
       "      <td>3IPTG</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>493008</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>528758</td>\n",
       "      <td>4.628800e-14</td>\n",
       "      <td>80</td>\n",
       "      <td>44</td>\n",
       "      <td>36</td>\n",
       "      <td>81.818184</td>\n",
       "      <td>AACGTTTTTTTTTTTTTTTTTGAAAGA</td>\n",
       "      <td>ACCCTTTTTTTTTTTTTTTTTGAAAGA</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN435</td>\n",
       "      <td>c.-3939_-3937delCGTinsGGG</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>6.354350</td>\n",
       "      <td>3IPTG</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN       FRAC  \\\n",
       "20557        p2_2_S1     2  629173  1.383230e-14   30    17    13  76.470589   \n",
       "33962        p2_2_S2     2  113969  2.790510e-07    6     4     2  50.000000   \n",
       "35494        p2_2_S2     2  141270  4.924100e-14   59    40    19  47.500000   \n",
       "57248        p2_2_S2     2  528761  2.688110e-14   45    33    12  36.363636   \n",
       "63072        p2_2_S2     2  629170  2.953720e-15   57    38    19  50.000000   \n",
       "85859        p2_2_S3     2  141267  1.219870e-13  159    97    62  63.917526   \n",
       "161975       p3_D_S2     3  195424  4.551630e-13  413   239   174  72.803345   \n",
       "214268       p3_D_S2     3  629169  2.384300e-13  355   212   143  67.452827   \n",
       "274642       p3_D_S3     3  528755  8.763400e-14  140    84    56  66.666664   \n",
       "280252       p3_D_S3     3  618608  8.492520e-09    7     5     2  40.000000   \n",
       "406624   p3IPTG_A_S1    18  141270  2.114290e-13  235   158    77  48.734177   \n",
       "447613   p3IPTG_A_S1    18  618608  2.102690e-14    9     7     2  28.571428   \n",
       "448420   p3IPTG_A_S1    18  629167  7.256610e-14  246   139   107  76.978416   \n",
       "493008  p3IPTG_A_S1b    18  528758  4.628800e-14   80    44    36  81.818184   \n",
       "\n",
       "                                              REF  \\\n",
       "20557                  GTTTTTTTTTTTTTTTTAGTTTGAAC   \n",
       "33962                                           T   \n",
       "35494                      CAGAGAGAGAGAGAGAGAGAGC   \n",
       "57248                    GTTTTTTTTTTTTTTTTTGAAAGA   \n",
       "63072               CTAGTTTTTTTTTTTTTTTTAGTTTGAAC   \n",
       "85859                   TCTCAGAGAGAGAGAGAGAGAGAGC   \n",
       "161975        TTTCCAAAAAAAAAAAAAAAGTAAAATAGAAAAGC   \n",
       "214268             TCTAGTTTTTTTTTTTTTTTTAGTTTGAAC   \n",
       "274642  TCAAACGTTTTTTTTTTTTTTTTTGAAAGAAATTGATTGCT   \n",
       "280252                                          A   \n",
       "406624                     CAGAGAGAGAGAGAGAGAGAGC   \n",
       "447613                                          A   \n",
       "448420     ATTCTAGTTTTTTTTTTTTTTTTAGTTTGAACTCAAAA   \n",
       "493008                AACGTTTTTTTTTTTTTTTTTGAAAGA   \n",
       "\n",
       "                                              ALT                    EFF  \\\n",
       "20557                  GTTTTTTTTTATTTATAAGTTTGAAC  upstream_gene_variant   \n",
       "33962                                           G  upstream_gene_variant   \n",
       "35494                        CAGAGAGAGAGAGAGAGAGC  upstream_gene_variant   \n",
       "57248                     GTTTTTTTTTTTTTTTTGAAAGA  upstream_gene_variant   \n",
       "63072              TTAGATTTTTTTTTTTTTTTTAGTTTGAAC  upstream_gene_variant   \n",
       "85859                 ACTCAGAGAGAGAGAGAGAGAGAGAGC  upstream_gene_variant   \n",
       "161975      ATTCTGAAAAAAAAAAAAAAAAGTAAAATAGAAAAGC  upstream_gene_variant   \n",
       "214268             TCTAGTTTTTTTTTTTTCTATAGTTTGAAC  upstream_gene_variant   \n",
       "274642  CCAATCGTTTTTTTTTTTTTTTTTGAAAGAAATTGATTGCT  upstream_gene_variant   \n",
       "280252                                          G  upstream_gene_variant   \n",
       "406624                     CAGAGAGAAAAAGAGAGAGAGC  upstream_gene_variant   \n",
       "447613                                          G  upstream_gene_variant   \n",
       "448420       TTTTTAATTTTTTTTTTTTTTAGTTTGAACTCAAAA  upstream_gene_variant   \n",
       "493008                ACCCTTTTTTTTTTTTTTTTTGAAAGA  upstream_gene_variant   \n",
       "\n",
       "          IMPACT     AFF                                                MUT  \\\n",
       "20557   MODIFIER  MPN508                  c.-2605_-2599delAAAAAAAinsTATAAAT   \n",
       "33962   MODIFIER  MPN094                                         c.-2292T>G   \n",
       "35494   MODIFIER  MPN109                                   c.-624_-623delAG   \n",
       "57248   MODIFIER  MPN435                                        c.-3956delA   \n",
       "63072   MODIFIER  MPN508  c.-2614_-2586delGTTCAAACTAAAAAAAAAAAAAAAACTAGi...   \n",
       "85859   MODIFIER  MPN109  c.-646_-622delTCTCAGAGAGAGAGAGAGAGAGAGCinsACTC...   \n",
       "161975  MODIFIER  MPN148  c.-97_-63delTTTCCAAAAAAAAAAAAAAAGTAAAATAGAAAAG...   \n",
       "214268  MODIFIER  MPN508                          c.-2604_-2602delAAAinsTAG   \n",
       "274642  MODIFIER  MPN435                      c.-3937_-3933delTTTGAinsATTGG   \n",
       "280252  MODIFIER  MPN504                                          c.-702A>G   \n",
       "406624  MODIFIER  MPN109                            c.-635_-633delGAGinsAAA   \n",
       "447613  MODIFIER  MPN504                                          c.-702A>G   \n",
       "448420  MODIFIER  MPN508  c.-2620_-2583delTTTTGAGTTCAAACTAAAAAAAAAAAAAAA...   \n",
       "493008  MODIFIER  MPN435                          c.-3939_-3937delCGTinsGGG   \n",
       "\n",
       "          ANN_TYPE  log2Frac Condition  annotation  \n",
       "20557   intergenic  6.256833         2  intergenic  \n",
       "33962   intergenic  5.643856         2  intergenic  \n",
       "35494   intergenic  5.569856         2  intergenic  \n",
       "57248   intergenic  5.184424         2  intergenic  \n",
       "63072   intergenic  5.643856         2  intergenic  \n",
       "85859   intergenic  5.998140         2  intergenic  \n",
       "161975  intergenic  6.185933         3  intergenic  \n",
       "214268  intergenic  6.075807         3  intergenic  \n",
       "274642  intergenic  6.058894         3  intergenic  \n",
       "280252  intergenic  5.321928         3  intergenic  \n",
       "406624  intergenic  5.606862     3IPTG  intergenic  \n",
       "447613  intergenic  4.836501     3IPTG  intergenic  \n",
       "448420  intergenic  6.266382     3IPTG  intergenic  \n",
       "493008  intergenic  6.354350     3IPTG  intergenic  "
      ]
     },
     "execution_count": 635,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subdf1[(subdf1['FRAC']>=25) & (subdf1['ANN_TYPE']=='intergenic')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 632,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c432e6b97ba44116b4c50fec3cdf243c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0deae078c69047fe85b15fd9e47ba75c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(array([0, 1]), <a list of 2 Text xticklabel objects>)"
      ]
     },
     "execution_count": 632,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For genes\n",
    "plt.figure()\n",
    "plt.subplot(2,1,1)\n",
    "sns.swarmplot(x='IMPACT', y='FRAC', hue='PASS', data=subdf1[(subdf1['FRAC']>=5) & (subdf1['ANN_TYPE']=='gene')])\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "\n",
    "plt.subplot(2,1,2)\n",
    "sns.swarmplot(x='IMPACT', y='FRAC', hue='PASS',data=subdf2[(subdf2['FRAC']>=5) & (subdf2['ANN_TYPE']=='gene')])\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "\n",
    "\n",
    "# For intergenic\n",
    "plt.figure()\n",
    "plt.subplot(2,1,1)\n",
    "sns.swarmplot(x='IMPACT', y='FRAC', hue='PASS',data=subdf1[(subdf1['FRAC']>=5) & (subdf1['ANN_TYPE']=='intergenic')])\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "\n",
    "plt.subplot(2,1,2)\n",
    "sns.swarmplot(x='IMPACT', y='FRAC', hue='PASS',data=subdf2[(subdf2['FRAC']>=5) & (subdf2['ANN_TYPE']=='intergenic')])\n",
    "plt.xticks(rotation=45, ha='right')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_fixation(df):\n",
    "    \"\"\" Plot to show the percentage of variants mapping to each type of annotation \"\"\"\n",
    "    colord = {'intergenic':0, 'gene':1, 'essential':2, 'non-essential':3, 'cassette':4}\n",
    "    pasind = {2:0, 3:1, 15:2, 18:3}\n",
    "    rs = {}\n",
    "    colors = []\n",
    "    for k, v in Counter(list(df['POS'])).items():\n",
    "        if v>2:\n",
    "            rs[k] = [0.0,0.0,0.0,0.0]\n",
    "            subdf = df[df['POS']==k]\n",
    "            for passage in set(subdf['PASS']):\n",
    "                rs[k][pasind[passage]] = list(subdf[subdf['PASS']==passage]['FRAC'])\n",
    "    return rs\n",
    "dic  = plot_fixation(data1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p2_2_S1</th>\n",
       "      <th>p2_2_S2</th>\n",
       "      <th>p2_2_S3</th>\n",
       "      <th>p3_D_S2</th>\n",
       "      <th>p3_D_S3</th>\n",
       "      <th>p15_F_S1</th>\n",
       "      <th>p15_F_S5</th>\n",
       "      <th>p3IPTG_A_S1</th>\n",
       "      <th>p3IPTG_A_S1b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>p2_2_S2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.304095</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>p2_2_S3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.499998</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.368914</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.297108</td>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.499997</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.481911</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>p3_D_S3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.368914</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.499996</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.370263</td>\n",
       "      <td>0.000963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>p15_F_S1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.304095</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>p15_F_S5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.481911</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.499997</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.297108</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.370263</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.499991</td>\n",
       "      <td>0.008616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008616</td>\n",
       "      <td>0.499999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              p2_2_S1   p2_2_S2   p2_2_S3   p3_D_S2   p3_D_S3  p15_F_S1  \\\n",
       "p2_2_S1           0.5  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "p2_2_S2           0.0  0.500000  0.000000  0.000000  0.000000  0.304095   \n",
       "p2_2_S3           0.0  0.000000  0.499998  0.000000  0.368914  0.000000   \n",
       "p3_D_S2           0.0  0.000000  0.000000  0.499997  0.000000  0.000000   \n",
       "p3_D_S3           0.0  0.000000  0.368914  0.000000  0.499996  0.000000   \n",
       "p15_F_S1          0.0  0.304095  0.000000  0.000000  0.000000  0.500000   \n",
       "p15_F_S5          0.0  0.000000  0.000000  0.481911  0.000000  0.000000   \n",
       "p3IPTG_A_S1       0.0  0.000000  0.297108  0.000000  0.370263  0.000000   \n",
       "p3IPTG_A_S1b      0.0  0.000000  0.000005  0.000000  0.000963  0.000000   \n",
       "\n",
       "              p15_F_S5  p3IPTG_A_S1  p3IPTG_A_S1b  \n",
       "p2_2_S1       0.000000     0.000000      0.000000  \n",
       "p2_2_S2       0.000000     0.000000      0.000000  \n",
       "p2_2_S3       0.000000     0.297108      0.000005  \n",
       "p3_D_S2       0.481911     0.000000      0.000000  \n",
       "p3_D_S3       0.000000     0.370263      0.000963  \n",
       "p15_F_S1      0.000000     0.000000      0.000000  \n",
       "p15_F_S5      0.499997     0.000000      0.000000  \n",
       "p3IPTG_A_S1   0.000000     0.499991      0.008616  \n",
       "p3IPTG_A_S1b  0.000000     0.008616      0.499999  "
      ]
     },
     "execution_count": 327,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rs = {}\n",
    "columns = ['p2_2_S1', 'p2_2_S2', 'p2_2_S3', 'p3_D_S2', 'p3_D_S3', 'p15_F_S1', 'p15_F_S5',  'p3IPTG_A_S1', 'p3IPTG_A_S1b']\n",
    "for n in columns:\n",
    "    rs[n] = []\n",
    "    for m in columns:\n",
    "        rs[n].append(round(mannwhitneyu(data1[(data1['SAMPLE']==n) & (data1['FRAC']>=1)]['FRAC'], data1[(data1['SAMPLE']==m) & (data1['FRAC']>=1)]['FRAC'])[1], 8))\n",
    "pvalues = pd.DataFrame.from_dict(rs, orient='index')\n",
    "pvalues.columns=columns\n",
    "pvalues\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extracta(df, sample):\n",
    "    rs = []\n",
    "    subdf = df[df['SAMPLE']==sample].copy()\n",
    "    fractions = {}\n",
    "    for k, v in zip(subdf['POS'], subdf['FRAC']):\n",
    "        if k in fractions:\n",
    "            fractions[k].append(v)\n",
    "        else:\n",
    "            fractions[k] = [v]\n",
    "    fractions = {k:np.mean(v) for k, v in fractions.items()}\n",
    "    for i in range(1, len(genome_C5)+1):\n",
    "        if i in fractions:\n",
    "            if fractions[i]>=1:\n",
    "                rs.append(i)\n",
    "            else:\n",
    "                rs.append(1)\n",
    "    return np.array(rs)\n",
    "\n",
    "def volcano():\n",
    "    rs = {}\n",
    "    for sample in set(data1[data1['PASS'].isin([3,18])]['SAMPLE']):\n",
    "         rs[sample] = extracta(data1, sample)\n",
    "    fcs = []\n",
    "    pvals = []\n",
    "    positions = []\n",
    "    for i in range(1, len(genome_C5)+1):\n",
    "        try:\n",
    "            vals = np.array([rs['p3IPTG_A_S1b'][i], rs['p3IPTG_A_S1'][i], rs['p3_D_S3'][i], rs['p3_D_S2'][i]])\n",
    "            if len(vals[vals>1])==0:\n",
    "                pass\n",
    "            else:\n",
    "                fc = np.log2(np.mean([rs['p3IPTG_A_S1b'][i], rs['p3IPTG_A_S1'][i]]))-np.log2(np.mean([rs['p3_D_S3'][i], rs['p3_D_S2'][i]]))\n",
    "                pval = -1*np.log10(mannwhitneyu([rs['p3IPTG_A_S1b'][i], rs['p3IPTG_A_S1'][i]], [rs['p3_D_S3'][i], rs['p3_D_S2'][i]])[1])\n",
    "                fcs.append(fc)\n",
    "                pvals.append(pval)\n",
    "                positions.append(i)\n",
    "        except:\n",
    "            pass\n",
    "    return fcs, pvals, positions\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {},
   "outputs": [],
   "source": [
    "f, pv, p  = volcano()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/smiravet/.local/lib/python3.6/site-packages/ipykernel_launcher.py:1: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ff30311529494c1380444d9470116f74",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.lines.Line2D at 0x7fef2ff662b0>"
      ]
     },
     "execution_count": 357,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plt.figure()\n",
    "plt.scatter(f, pv)\n",
    "plt.axhline(y=-1*np.log10(0.05), linestyle='--', color='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 601,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/smiravet/.local/lib/python3.6/site-packages/ipykernel_launcher.py:2: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  \n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1148eeaa2592464383b4485d207b3de9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "background = np.log2(data1[(data1['PASS']==3) & (data1['FRAC']>=1)]['FRAC'])\n",
    "plt.figure()\n",
    "sns.distplot(background)\n",
    "plt.axvline(x=np.mean(background))\n",
    "plt.axvline(x=np.percentile(background,99), color='r', linestyle='--')\n",
    "selected = data1[data1['PASS']==18].copy()\n",
    "#selected['pvalue'] = [mannwhitneyu(frac, background) for frac in list(selected['FRAC'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7ac974ea6c9e431a8b3f1d53bde7b566",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure()\n",
    "c = 1\n",
    "for pas in [2, 3, 15, 18]:\n",
    "    plt.subplot(4,1, c)\n",
    "    sns.scatterplot(x='POS', y='FRAC', hue='IMPACT', data=data1[(data1['FRAC']>2.5) & (data1['PASS']==pas) & (~data1['ANN_TYPE'].isin(['essential', 'non-essential']))])\n",
    "    c+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>405668</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>130837</td>\n",
       "      <td>6.795500e-15</td>\n",
       "      <td>41</td>\n",
       "      <td>39</td>\n",
       "      <td>2</td>\n",
       "      <td>5.128205</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN101</td>\n",
       "      <td>p.Ser133Gly</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>496576</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>602108</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>40</td>\n",
       "      <td>38</td>\n",
       "      <td>2</td>\n",
       "      <td>5.263158</td>\n",
       "      <td>CAACTGTAA</td>\n",
       "      <td>TAACAGTAA</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN489</td>\n",
       "      <td>p.Gln254Leu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>496332</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>595882</td>\n",
       "      <td>4.776910e-15</td>\n",
       "      <td>20</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>5.263158</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN485</td>\n",
       "      <td>p.Gly389Arg</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>446239</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>602189</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>277</td>\n",
       "      <td>263</td>\n",
       "      <td>14</td>\n",
       "      <td>5.323194</td>\n",
       "      <td>TTGGGTGAGGGGGAGGGGGTTT</td>\n",
       "      <td>TAGGGAGAGGGGGAGGGGGTTT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN489</td>\n",
       "      <td>p.ThrGln227SerLeu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>481872</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>284317</td>\n",
       "      <td>5.728050e-16</td>\n",
       "      <td>158</td>\n",
       "      <td>150</td>\n",
       "      <td>8</td>\n",
       "      <td>5.333333</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>MPN233</td>\n",
       "      <td>p.Leu418Leu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>412492</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>207198</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>291</td>\n",
       "      <td>276</td>\n",
       "      <td>15</td>\n",
       "      <td>5.434783</td>\n",
       "      <td>AGTACCACCACCA</td>\n",
       "      <td>GGGACCTCCGCCG</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN154</td>\n",
       "      <td>p.Val478Gly</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>446222</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>602038</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>58</td>\n",
       "      <td>55</td>\n",
       "      <td>3</td>\n",
       "      <td>5.454545</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN489</td>\n",
       "      <td>p.Thr279Ser</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>489010</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>442873</td>\n",
       "      <td>1.190810e-15</td>\n",
       "      <td>115</td>\n",
       "      <td>109</td>\n",
       "      <td>6</td>\n",
       "      <td>5.504587</td>\n",
       "      <td>GTGCTTAGTGGT</td>\n",
       "      <td>GAGCTAAGAGGA</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>MPN370</td>\n",
       "      <td>p.CysLeuValVal542*</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>484509</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>340718</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>5.555555</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>MPN285</td>\n",
       "      <td>p.Leu170Leu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>496334</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>596162</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>5.555555</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>MPN485</td>\n",
       "      <td>p.Thr295Thr</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>410476</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>184026</td>\n",
       "      <td>1.371150e-15</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>5.555555</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN141</td>\n",
       "      <td>p.Gly1067Glu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>477062</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>187323</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>5.555555</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>MPN142</td>\n",
       "      <td>p.Gly536Gly</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>476267</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>170557</td>\n",
       "      <td>1.003550e-14</td>\n",
       "      <td>132</td>\n",
       "      <td>125</td>\n",
       "      <td>7</td>\n",
       "      <td>5.600000</td>\n",
       "      <td>ATTCACCAACAAGAA</td>\n",
       "      <td>TTTTACCAACAAGAA</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN131</td>\n",
       "      <td>p.HisSer174LeuLeu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>479961</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>244781</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>5.882353</td>\n",
       "      <td>GTGTTTTTAAAAATAG</td>\n",
       "      <td>GG</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>MPN201</td>\n",
       "      <td>p.Val112fs</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>480152</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>249385</td>\n",
       "      <td>3.668960e-15</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>5.882353</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>MPN205</td>\n",
       "      <td>p.Tyr286*</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>496333</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>595885</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>5.882353</td>\n",
       "      <td>GACC</td>\n",
       "      <td>TACT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN485</td>\n",
       "      <td>p.GlyGln387SerLys</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>446232</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>602108</td>\n",
       "      <td>2.877960e-14</td>\n",
       "      <td>85</td>\n",
       "      <td>80</td>\n",
       "      <td>5</td>\n",
       "      <td>6.250000</td>\n",
       "      <td>CAACTGTAAGGCACTACCAAT</td>\n",
       "      <td>CAACTGTAAGGCCCTCCCACT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN489</td>\n",
       "      <td>p.IleGlySer249SerGlyArg</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>495289</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>572821</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN463</td>\n",
       "      <td>p.Lys43Arg</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>480151</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>249382</td>\n",
       "      <td>4.558190e-15</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>MPN205</td>\n",
       "      <td>p.Leu285Leu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>446344</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>603384</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>232</td>\n",
       "      <td>217</td>\n",
       "      <td>15</td>\n",
       "      <td>6.912442</td>\n",
       "      <td>TTGGCTAGCTCGCT</td>\n",
       "      <td>TTGGCTAGCTCACA</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN490</td>\n",
       "      <td>p.Ala335Val</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>437138</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>497363</td>\n",
       "      <td>2.787740e-15</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>7.142857</td>\n",
       "      <td>AACA</td>\n",
       "      <td>TACT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN412</td>\n",
       "      <td>p.AsnIle259TyrPhe</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>488940</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>441800</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>58</td>\n",
       "      <td>54</td>\n",
       "      <td>4</td>\n",
       "      <td>7.407407</td>\n",
       "      <td>GACGGCGCGCCCCAG</td>\n",
       "      <td>AACAGCGTGCCCCAG</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN370</td>\n",
       "      <td>p.AspGlyAla184AsnSerVal</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>501333</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>695356</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>214</td>\n",
       "      <td>199</td>\n",
       "      <td>15</td>\n",
       "      <td>7.537688</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN567</td>\n",
       "      <td>p.Glu557Lys</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>491482</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>497614</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>7.692307</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN413</td>\n",
       "      <td>p.Ser38Trp</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>488746</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>436072</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>8.333333</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN365</td>\n",
       "      <td>p.Leu167Pro</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>446224</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>602067</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>8.333333</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN489</td>\n",
       "      <td>p.Gln269Arg</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>496643</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>603403</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>91</td>\n",
       "      <td>84</td>\n",
       "      <td>7</td>\n",
       "      <td>8.333333</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>MPN490</td>\n",
       "      <td>p.Pro332Pro</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>494181</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>552779</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>51</td>\n",
       "      <td>47</td>\n",
       "      <td>4</td>\n",
       "      <td>8.510638</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>MPN453</td>\n",
       "      <td>p.Gly232Gly</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>411053</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>190954</td>\n",
       "      <td>6.336400e-15</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>9.090909</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN144</td>\n",
       "      <td>p.Tyr122Cys</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>437173</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>498194</td>\n",
       "      <td>6.336400e-15</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>9.090909</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN414</td>\n",
       "      <td>p.Asp55Ala</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>484510</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>340741</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>9.090909</td>\n",
       "      <td>GCT</td>\n",
       "      <td>ACA</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN285</td>\n",
       "      <td>p.Leu178Gln</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>446227</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>602075</td>\n",
       "      <td>6.117690e-15</td>\n",
       "      <td>35</td>\n",
       "      <td>32</td>\n",
       "      <td>3</td>\n",
       "      <td>9.375000</td>\n",
       "      <td>ATT</td>\n",
       "      <td>CTC</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN489</td>\n",
       "      <td>p.Asn266Glu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>476946</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>183951</td>\n",
       "      <td>1.514530e-15</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>11.111111</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN141</td>\n",
       "      <td>p.Tyr1042Phe</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>446223</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>602050</td>\n",
       "      <td>1.522060e-15</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>11.111111</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN489</td>\n",
       "      <td>p.Thr275Pro</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>495288</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>572799</td>\n",
       "      <td>3.819640e-15</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>12.500000</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN463</td>\n",
       "      <td>p.Arg36Cys</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>454539</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>695356</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>450</td>\n",
       "      <td>400</td>\n",
       "      <td>50</td>\n",
       "      <td>12.500000</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN567</td>\n",
       "      <td>p.Glu557Lys</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>437174</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>498234</td>\n",
       "      <td>1.140750e-12</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666666</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>MPN414</td>\n",
       "      <td>p.Asp68Asp</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>445798</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>596881</td>\n",
       "      <td>4.386870e-10</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN485</td>\n",
       "      <td>p.Thr56Pro</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>410351</td>\n",
       "      <td>p3IPTG_A_S1</td>\n",
       "      <td>18</td>\n",
       "      <td>182761</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>299</td>\n",
       "      <td>248</td>\n",
       "      <td>51</td>\n",
       "      <td>20.564516</td>\n",
       "      <td>AAGTAGTAGTAGTAGTAGTAGTC</td>\n",
       "      <td>AAGTAGTAGTAGTAGTAGTC</td>\n",
       "      <td>conservative_inframe_deletion</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN141</td>\n",
       "      <td>p.Ser652del</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>476893</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>182758</td>\n",
       "      <td>3.085870e-14</td>\n",
       "      <td>124</td>\n",
       "      <td>102</td>\n",
       "      <td>22</td>\n",
       "      <td>21.568628</td>\n",
       "      <td>TTGAAGTAGTAGTAGTAGTAGTAGTC</td>\n",
       "      <td>ATGAAGTAGTAGTAGTAGTAGTC</td>\n",
       "      <td>conservative_inframe_deletion&amp;synonymous_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN141</td>\n",
       "      <td>p.Ser646del</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>476841</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>181560</td>\n",
       "      <td>1.848900e-07</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN141</td>\n",
       "      <td>p.Glu245Gly</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN       FRAC  \\\n",
       "405668   p3IPTG_A_S1    18  130837  6.795500e-15   41    39     2   5.128205   \n",
       "496576  p3IPTG_A_S1b    18  602108  0.000000e+00   40    38     2   5.263158   \n",
       "496332  p3IPTG_A_S1b    18  595882  4.776910e-15   20    19     1   5.263158   \n",
       "446239   p3IPTG_A_S1    18  602189  0.000000e+00  277   263    14   5.323194   \n",
       "481872  p3IPTG_A_S1b    18  284317  5.728050e-16  158   150     8   5.333333   \n",
       "412492   p3IPTG_A_S1    18  207198  0.000000e+00  291   276    15   5.434783   \n",
       "446222   p3IPTG_A_S1    18  602038  0.000000e+00   58    55     3   5.454545   \n",
       "489010  p3IPTG_A_S1b    18  442873  1.190810e-15  115   109     6   5.504587   \n",
       "484509  p3IPTG_A_S1b    18  340718  0.000000e+00   19    18     1   5.555555   \n",
       "496334  p3IPTG_A_S1b    18  596162  0.000000e+00   19    18     1   5.555555   \n",
       "410476   p3IPTG_A_S1    18  184026  1.371150e-15   19    18     1   5.555555   \n",
       "477062  p3IPTG_A_S1b    18  187323  0.000000e+00   19    18     1   5.555555   \n",
       "476267  p3IPTG_A_S1b    18  170557  1.003550e-14  132   125     7   5.600000   \n",
       "479961  p3IPTG_A_S1b    18  244781  0.000000e+00   18    17     1   5.882353   \n",
       "480152  p3IPTG_A_S1b    18  249385  3.668960e-15   18    17     1   5.882353   \n",
       "496333  p3IPTG_A_S1b    18  595885  0.000000e+00   18    17     1   5.882353   \n",
       "446232   p3IPTG_A_S1    18  602108  2.877960e-14   85    80     5   6.250000   \n",
       "495289  p3IPTG_A_S1b    18  572821  0.000000e+00   16    15     1   6.666667   \n",
       "480151  p3IPTG_A_S1b    18  249382  4.558190e-15   16    15     1   6.666667   \n",
       "446344   p3IPTG_A_S1    18  603384  0.000000e+00  232   217    15   6.912442   \n",
       "437138   p3IPTG_A_S1    18  497363  2.787740e-15   15    14     1   7.142857   \n",
       "488940  p3IPTG_A_S1b    18  441800  0.000000e+00   58    54     4   7.407407   \n",
       "501333  p3IPTG_A_S1b    18  695356  0.000000e+00  214   199    15   7.537688   \n",
       "491482  p3IPTG_A_S1b    18  497614  0.000000e+00   14    13     1   7.692307   \n",
       "488746  p3IPTG_A_S1b    18  436072  0.000000e+00   13    12     1   8.333333   \n",
       "446224   p3IPTG_A_S1    18  602067  0.000000e+00   13    12     1   8.333333   \n",
       "496643  p3IPTG_A_S1b    18  603403  0.000000e+00   91    84     7   8.333333   \n",
       "494181  p3IPTG_A_S1b    18  552779  0.000000e+00   51    47     4   8.510638   \n",
       "411053   p3IPTG_A_S1    18  190954  6.336400e-15   12    11     1   9.090909   \n",
       "437173   p3IPTG_A_S1    18  498194  6.336400e-15   12    11     1   9.090909   \n",
       "484510  p3IPTG_A_S1b    18  340741  0.000000e+00   12    11     1   9.090909   \n",
       "446227   p3IPTG_A_S1    18  602075  6.117690e-15   35    32     3   9.375000   \n",
       "476946  p3IPTG_A_S1b    18  183951  1.514530e-15   10     9     1  11.111111   \n",
       "446223   p3IPTG_A_S1    18  602050  1.522060e-15   10     9     1  11.111111   \n",
       "495288  p3IPTG_A_S1b    18  572799  3.819640e-15    9     8     1  12.500000   \n",
       "454539   p3IPTG_A_S1    18  695356  0.000000e+00  450   400    50  12.500000   \n",
       "437174   p3IPTG_A_S1    18  498234  1.140750e-12    7     6     1  16.666666   \n",
       "445798   p3IPTG_A_S1    18  596881  4.386870e-10    6     5     1  20.000000   \n",
       "410351   p3IPTG_A_S1    18  182761  0.000000e+00  299   248    51  20.564516   \n",
       "476893  p3IPTG_A_S1b    18  182758  3.085870e-14  124   102    22  21.568628   \n",
       "476841  p3IPTG_A_S1b    18  181560  1.848900e-07    5     4     1  25.000000   \n",
       "\n",
       "                               REF                      ALT  \\\n",
       "405668                           A                        G   \n",
       "496576                   CAACTGTAA                TAACAGTAA   \n",
       "496332                           C                        T   \n",
       "446239      TTGGGTGAGGGGGAGGGGGTTT   TAGGGAGAGGGGGAGGGGGTTT   \n",
       "481872                           C                        T   \n",
       "412492               AGTACCACCACCA            GGGACCTCCGCCG   \n",
       "446222                           T                        A   \n",
       "489010                GTGCTTAGTGGT             GAGCTAAGAGGA   \n",
       "484509                           C                        T   \n",
       "496334                           T                        C   \n",
       "410476                           G                        A   \n",
       "477062                           G                        A   \n",
       "476267             ATTCACCAACAAGAA          TTTTACCAACAAGAA   \n",
       "479961            GTGTTTTTAAAAATAG                       GG   \n",
       "480152                           C                        G   \n",
       "496333                        GACC                     TACT   \n",
       "446232       CAACTGTAAGGCACTACCAAT    CAACTGTAAGGCCCTCCCACT   \n",
       "495289                           A                        G   \n",
       "480151                           G                        T   \n",
       "446344              TTGGCTAGCTCGCT           TTGGCTAGCTCACA   \n",
       "437138                        AACA                     TACT   \n",
       "488940             GACGGCGCGCCCCAG          AACAGCGTGCCCCAG   \n",
       "501333                           C                        T   \n",
       "491482                           C                        G   \n",
       "488746                           T                        C   \n",
       "446224                           T                        C   \n",
       "496643                           T                        C   \n",
       "494181                           G                        A   \n",
       "411053                           A                        G   \n",
       "437173                           A                        C   \n",
       "484510                         GCT                      ACA   \n",
       "446227                         ATT                      CTC   \n",
       "476946                           A                        T   \n",
       "446223                           T                        G   \n",
       "495288                           C                        T   \n",
       "454539                           C                        T   \n",
       "437174                           T                        C   \n",
       "445798                           T                        G   \n",
       "410351     AAGTAGTAGTAGTAGTAGTAGTC     AAGTAGTAGTAGTAGTAGTC   \n",
       "476893  TTGAAGTAGTAGTAGTAGTAGTAGTC  ATGAAGTAGTAGTAGTAGTAGTC   \n",
       "476841                           A                        G   \n",
       "\n",
       "                                                     EFF    IMPACT     AFF  \\\n",
       "405668                                  missense_variant  MODERATE  MPN101   \n",
       "496576                                  missense_variant  MODERATE  MPN489   \n",
       "496332                                  missense_variant  MODERATE  MPN485   \n",
       "446239                                  missense_variant  MODERATE  MPN489   \n",
       "481872                                synonymous_variant       LOW  MPN233   \n",
       "412492                                  missense_variant  MODERATE  MPN154   \n",
       "446222                                  missense_variant  MODERATE  MPN489   \n",
       "489010                                       stop_gained      HIGH  MPN370   \n",
       "484509                                synonymous_variant       LOW  MPN285   \n",
       "496334                                synonymous_variant       LOW  MPN485   \n",
       "410476                                  missense_variant  MODERATE  MPN141   \n",
       "477062                                synonymous_variant       LOW  MPN142   \n",
       "476267                                  missense_variant  MODERATE  MPN131   \n",
       "479961                                frameshift_variant      HIGH  MPN201   \n",
       "480152                                       stop_gained      HIGH  MPN205   \n",
       "496333                                  missense_variant  MODERATE  MPN485   \n",
       "446232                                  missense_variant  MODERATE  MPN489   \n",
       "495289                                  missense_variant  MODERATE  MPN463   \n",
       "480151                                synonymous_variant       LOW  MPN205   \n",
       "446344                                  missense_variant  MODERATE  MPN490   \n",
       "437138                                  missense_variant  MODERATE  MPN412   \n",
       "488940                                  missense_variant  MODERATE  MPN370   \n",
       "501333                                  missense_variant  MODERATE  MPN567   \n",
       "491482                                  missense_variant  MODERATE  MPN413   \n",
       "488746                                  missense_variant  MODERATE  MPN365   \n",
       "446224                                  missense_variant  MODERATE  MPN489   \n",
       "496643                                synonymous_variant       LOW  MPN490   \n",
       "494181                                synonymous_variant       LOW  MPN453   \n",
       "411053                                  missense_variant  MODERATE  MPN144   \n",
       "437173                                  missense_variant  MODERATE  MPN414   \n",
       "484510                                  missense_variant  MODERATE  MPN285   \n",
       "446227                                  missense_variant  MODERATE  MPN489   \n",
       "476946                                  missense_variant  MODERATE  MPN141   \n",
       "446223                                  missense_variant  MODERATE  MPN489   \n",
       "495288                                  missense_variant  MODERATE  MPN463   \n",
       "454539                                  missense_variant  MODERATE  MPN567   \n",
       "437174                                synonymous_variant       LOW  MPN414   \n",
       "445798                                  missense_variant  MODERATE  MPN485   \n",
       "410351                     conservative_inframe_deletion  MODERATE  MPN141   \n",
       "476893  conservative_inframe_deletion&synonymous_variant  MODERATE  MPN141   \n",
       "476841                                  missense_variant  MODERATE  MPN141   \n",
       "\n",
       "                            MUT ANN_TYPE  \n",
       "405668              p.Ser133Gly     gene  \n",
       "496576              p.Gln254Leu     gene  \n",
       "496332              p.Gly389Arg     gene  \n",
       "446239        p.ThrGln227SerLeu     gene  \n",
       "481872              p.Leu418Leu     gene  \n",
       "412492              p.Val478Gly     gene  \n",
       "446222              p.Thr279Ser     gene  \n",
       "489010       p.CysLeuValVal542*     gene  \n",
       "484509              p.Leu170Leu     gene  \n",
       "496334              p.Thr295Thr     gene  \n",
       "410476             p.Gly1067Glu     gene  \n",
       "477062              p.Gly536Gly     gene  \n",
       "476267        p.HisSer174LeuLeu     gene  \n",
       "479961               p.Val112fs     gene  \n",
       "480152                p.Tyr286*     gene  \n",
       "496333        p.GlyGln387SerLys     gene  \n",
       "446232  p.IleGlySer249SerGlyArg     gene  \n",
       "495289               p.Lys43Arg     gene  \n",
       "480151              p.Leu285Leu     gene  \n",
       "446344              p.Ala335Val     gene  \n",
       "437138        p.AsnIle259TyrPhe     gene  \n",
       "488940  p.AspGlyAla184AsnSerVal     gene  \n",
       "501333              p.Glu557Lys     gene  \n",
       "491482               p.Ser38Trp     gene  \n",
       "488746              p.Leu167Pro     gene  \n",
       "446224              p.Gln269Arg     gene  \n",
       "496643              p.Pro332Pro     gene  \n",
       "494181              p.Gly232Gly     gene  \n",
       "411053              p.Tyr122Cys     gene  \n",
       "437173               p.Asp55Ala     gene  \n",
       "484510              p.Leu178Gln     gene  \n",
       "446227              p.Asn266Glu     gene  \n",
       "476946             p.Tyr1042Phe     gene  \n",
       "446223              p.Thr275Pro     gene  \n",
       "495288               p.Arg36Cys     gene  \n",
       "454539              p.Glu557Lys     gene  \n",
       "437174               p.Asp68Asp     gene  \n",
       "445798               p.Thr56Pro     gene  \n",
       "410351              p.Ser652del     gene  \n",
       "476893              p.Ser646del     gene  \n",
       "476841              p.Glu245Gly     gene  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data1[(data1['PASS']==18) & (data1['ANN_TYPE']!='cassette') & (data1['FRAC']>5)  & (data1['IMPACT']!='MODIFIER')].sort_values('FRAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_dic = {}\n",
    "for k, v in dic.items():\n",
    "    a = np.array(v)\n",
    "    try:\n",
    "        if len(a[1])==2 and len(a[-1])==2:\n",
    "            new_dic[k] = a\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [],
   "source": [
    "x3, x3i, x15 = [[], [], []]\n",
    "positions = []\n",
    "for k, v in dic.items():\n",
    "    a = np.array(v)\n",
    "    try:\n",
    "        if len(a[0])==2:\n",
    "            if len(a[1])==2:\n",
    "                x3.append(np.mean(a[1])/np.mean(a[0]))\n",
    "            elif len(a[2])==2:\n",
    "                x15.append(np.mean(a[2])/np.mean(a[0]))\n",
    "            elif len(a[-1])==2:\n",
    "                positions.append(k)\n",
    "                x3i.append(np.mean(a[-1])/np.mean(a[0]))\n",
    "            else:\n",
    "                pass\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "baed398c47f747c0b5381a6c1651a108",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.lines.Line2D at 0x7fef462b1e10>"
      ]
     },
     "execution_count": 271,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plt.figure()\n",
    "plt.violinplot([x3, x15, x3i])\n",
    "plt.xticks([1,2,3],['P3/P2', 'P15/P2', 'P3-IPTG/P2'])\n",
    "plt.axhline(y=np.mean(x3)+2*np.std(x3), color='red', linestyle='--', label='95% threshold P3/P2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0411993871078709"
      ]
     },
     "execution_count": 274,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thr = np.mean(x3)+2*np.std(x3)\n",
    "thr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = np.array(x3i)\n",
    "b = np.array(positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([211633, 264869, 198159, 224676, 254989, 458314, 553466, 675401,\n",
       "       767918, 807060])"
      ]
     },
     "execution_count": 282,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b[a>=thr]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Fixation of the mutations \n",
    "\n",
    "In this section we evaluate which mutations increase their representation from passage 2 to 3 and comparing this ratio between non-treated and induced with IPTG samples. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "      <th>ANN_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>137985</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>6772</td>\n",
       "      <td>1.989160e-15</td>\n",
       "      <td>582</td>\n",
       "      <td>569</td>\n",
       "      <td>13</td>\n",
       "      <td>2.284710</td>\n",
       "      <td>TCTTTAAGGAAAA</td>\n",
       "      <td>ACTTAAAGGAAAA</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN004</td>\n",
       "      <td>p.LeuPhe651HisLeu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>138227</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>8563</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>523</td>\n",
       "      <td>512</td>\n",
       "      <td>11</td>\n",
       "      <td>2.148438</td>\n",
       "      <td>AAACCAAAATAAACCAATGAAACAAGGCGTGT</td>\n",
       "      <td>AAACCAAAATAATCTATTGAAACAAGGCGTGT</td>\n",
       "      <td>initiator_codon_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>MPN006</td>\n",
       "      <td>p.Met1?</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>139095</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>15285</td>\n",
       "      <td>4.221290e-16</td>\n",
       "      <td>684</td>\n",
       "      <td>673</td>\n",
       "      <td>11</td>\n",
       "      <td>1.634472</td>\n",
       "      <td>CCCTAAAGAACAAGG</td>\n",
       "      <td>TCCTGAAGAACAAGG</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN013</td>\n",
       "      <td>p.Lys100Glu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>139787</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>20319</td>\n",
       "      <td>2.260140e-14</td>\n",
       "      <td>599</td>\n",
       "      <td>588</td>\n",
       "      <td>11</td>\n",
       "      <td>1.870748</td>\n",
       "      <td>CCCGTTCCCGCGTG</td>\n",
       "      <td>CCCGTTCCCGAGAG</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN018</td>\n",
       "      <td>p.Val336Glu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>140828</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>28083</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>644</td>\n",
       "      <td>633</td>\n",
       "      <td>11</td>\n",
       "      <td>1.737757</td>\n",
       "      <td>AATCCACGGCCGTTT</td>\n",
       "      <td>AATCCACGGCTGTCT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN022</td>\n",
       "      <td>p.ArgPhe260CysLeu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>267794</td>\n",
       "      <td>p3_D_S3</td>\n",
       "      <td>3</td>\n",
       "      <td>414743</td>\n",
       "      <td>2.170210e-15</td>\n",
       "      <td>172</td>\n",
       "      <td>161</td>\n",
       "      <td>11</td>\n",
       "      <td>6.832298</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>MPN347</td>\n",
       "      <td>p.Gly268Gly</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>269396</td>\n",
       "      <td>p3_D_S3</td>\n",
       "      <td>3</td>\n",
       "      <td>442878</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>235</td>\n",
       "      <td>222</td>\n",
       "      <td>13</td>\n",
       "      <td>5.855856</td>\n",
       "      <td>TAGTGGT</td>\n",
       "      <td>TGGAGGT</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>MPN370</td>\n",
       "      <td>p.Val544Glu</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>274642</td>\n",
       "      <td>p3_D_S3</td>\n",
       "      <td>3</td>\n",
       "      <td>528755</td>\n",
       "      <td>8.763400e-14</td>\n",
       "      <td>140</td>\n",
       "      <td>84</td>\n",
       "      <td>56</td>\n",
       "      <td>66.666664</td>\n",
       "      <td>TCAAACGTTTTTTTTTTTTTTTTTGAAAGAAATTGATTGCT</td>\n",
       "      <td>CCAATCGTTTTTTTTTTTTTTTTTGAAAGAAATTGATTGCT</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN435</td>\n",
       "      <td>c.-3937_-3933delTTTGAinsATTGG</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>280845</td>\n",
       "      <td>p3_D_S3</td>\n",
       "      <td>3</td>\n",
       "      <td>629170</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>175</td>\n",
       "      <td>104</td>\n",
       "      <td>71</td>\n",
       "      <td>68.269234</td>\n",
       "      <td>CTAGTTTTTTTTTTTTTTTTAGTTTGAACTCA</td>\n",
       "      <td>TTAGCTTTTTTTTTTTTTTTTAGTTTGAACTCA</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN508</td>\n",
       "      <td>c.-2617_-2586delTGAGTTCAAACTAAAAAAAAAAAAAAAACT...</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>286599</td>\n",
       "      <td>p3_D_S3</td>\n",
       "      <td>3</td>\n",
       "      <td>712979</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>228</td>\n",
       "      <td>216</td>\n",
       "      <td>12</td>\n",
       "      <td>5.555555</td>\n",
       "      <td>ACCCCCCCCGCTA</td>\n",
       "      <td>ACCCCCCCGCTA</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MPN579</td>\n",
       "      <td>c.-4224delG</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         SAMPLE  PASS     POS          QUAL  TOT  REFN  ALTN       FRAC  \\\n",
       "137985  p3_D_S2     3    6772  1.989160e-15  582   569    13   2.284710   \n",
       "138227  p3_D_S2     3    8563  0.000000e+00  523   512    11   2.148438   \n",
       "139095  p3_D_S2     3   15285  4.221290e-16  684   673    11   1.634472   \n",
       "139787  p3_D_S2     3   20319  2.260140e-14  599   588    11   1.870748   \n",
       "140828  p3_D_S2     3   28083  0.000000e+00  644   633    11   1.737757   \n",
       "...         ...   ...     ...           ...  ...   ...   ...        ...   \n",
       "267794  p3_D_S3     3  414743  2.170210e-15  172   161    11   6.832298   \n",
       "269396  p3_D_S3     3  442878  0.000000e+00  235   222    13   5.855856   \n",
       "274642  p3_D_S3     3  528755  8.763400e-14  140    84    56  66.666664   \n",
       "280845  p3_D_S3     3  629170  0.000000e+00  175   104    71  68.269234   \n",
       "286599  p3_D_S3     3  712979  0.000000e+00  228   216    12   5.555555   \n",
       "\n",
       "                                              REF  \\\n",
       "137985                              TCTTTAAGGAAAA   \n",
       "138227           AAACCAAAATAAACCAATGAAACAAGGCGTGT   \n",
       "139095                            CCCTAAAGAACAAGG   \n",
       "139787                             CCCGTTCCCGCGTG   \n",
       "140828                            AATCCACGGCCGTTT   \n",
       "...                                           ...   \n",
       "267794                                          T   \n",
       "269396                                    TAGTGGT   \n",
       "274642  TCAAACGTTTTTTTTTTTTTTTTTGAAAGAAATTGATTGCT   \n",
       "280845           CTAGTTTTTTTTTTTTTTTTAGTTTGAACTCA   \n",
       "286599                              ACCCCCCCCGCTA   \n",
       "\n",
       "                                              ALT                      EFF  \\\n",
       "137985                              ACTTAAAGGAAAA         missense_variant   \n",
       "138227           AAACCAAAATAATCTATTGAAACAAGGCGTGT  initiator_codon_variant   \n",
       "139095                            TCCTGAAGAACAAGG         missense_variant   \n",
       "139787                             CCCGTTCCCGAGAG         missense_variant   \n",
       "140828                            AATCCACGGCTGTCT         missense_variant   \n",
       "...                                           ...                      ...   \n",
       "267794                                          A       synonymous_variant   \n",
       "269396                                    TGGAGGT         missense_variant   \n",
       "274642  CCAATCGTTTTTTTTTTTTTTTTTGAAAGAAATTGATTGCT    upstream_gene_variant   \n",
       "280845          TTAGCTTTTTTTTTTTTTTTTAGTTTGAACTCA    upstream_gene_variant   \n",
       "286599                               ACCCCCCCGCTA    upstream_gene_variant   \n",
       "\n",
       "          IMPACT     AFF                                                MUT  \\\n",
       "137985  MODERATE  MPN004                                  p.LeuPhe651HisLeu   \n",
       "138227       LOW  MPN006                                            p.Met1?   \n",
       "139095  MODERATE  MPN013                                        p.Lys100Glu   \n",
       "139787  MODERATE  MPN018                                        p.Val336Glu   \n",
       "140828  MODERATE  MPN022                                  p.ArgPhe260CysLeu   \n",
       "...          ...     ...                                                ...   \n",
       "267794       LOW  MPN347                                        p.Gly268Gly   \n",
       "269396  MODERATE  MPN370                                        p.Val544Glu   \n",
       "274642  MODIFIER  MPN435                      c.-3937_-3933delTTTGAinsATTGG   \n",
       "280845  MODIFIER  MPN508  c.-2617_-2586delTGAGTTCAAACTAAAAAAAAAAAAAAAACT...   \n",
       "286599  MODIFIER  MPN579                                        c.-4224delG   \n",
       "\n",
       "          ANN_TYPE  \n",
       "137985        gene  \n",
       "138227        gene  \n",
       "139095        gene  \n",
       "139787        gene  \n",
       "140828        gene  \n",
       "...            ...  \n",
       "267794        gene  \n",
       "269396        gene  \n",
       "274642  intergenic  \n",
       "280845  intergenic  \n",
       "286599  intergenic  \n",
       "\n",
       "[68 rows x 15 columns]"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data1[(data1['ALTN']>10) & (data1['PASS']==3)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7cc32ecb3d864758ba1dccc70035c2b3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fa3e0a23a90>"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_fixation(df):\n",
    "    \n",
    "    subdf = df.groupby(['PASS', 'POS', 'ALT']).sum()\n",
    "    return subdf\n",
    "\n",
    "    positions = set([k for k, v in Counter(subdf['POS']).items() if v>=3])\n",
    "    return df[df['POS'].isin(positions)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['2', '6', 'T'],\n",
       "       ['2', '13', 'C'],\n",
       "       ['2', '31', 'A'],\n",
       "       ...,\n",
       "       ['18', '822913', 'TGCGTGAAA'],\n",
       "       ['18', '822925', 'C'],\n",
       "       ['18', '822931', 'CAAGTTCT']], dtype='<U73')"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.array([i for i in plot_fixation(snpcalls1).index])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "41542d7e65264784b0e4649fb3f6c70c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f1bd89350f0>"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plt.figure(figsize=(10,10))\n",
    "sns.scatterplot(x='POS', y='FRAC', hue='PASS', data=effect1[(effect1['IMPACT']=='HIGH') & (effect1['FRAC']>=1)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/smiravet/.local/lib/python3.6/site-packages/ipykernel_launcher.py:1: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3bfe1f931ce8464f8512a8a6005ef7b1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f92f451a828>"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plt.figure()\n",
    "effect21['logFRAC'] = np.log2(effect21['FRAC'])\n",
    "sns.violinplot(x='IMPACT', y='FRAC', hue='PASS', data=effect1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>ALTS</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>P</th>\n",
       "      <th>EFF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>18482</td>\n",
       "      <td>566325</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.469330e-15</td>\n",
       "      <td>205</td>\n",
       "      <td>204</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.487805</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>T|synonymous_variant|LOW|Gene_565549_566677|Ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18481</td>\n",
       "      <td>566311</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>7.046470e-16</td>\n",
       "      <td>198</td>\n",
       "      <td>197</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.505051</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>G|missense_variant|MODERATE|Gene_565549_566677...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18485</td>\n",
       "      <td>566383</td>\n",
       "      <td>GCC</td>\n",
       "      <td>CCT</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>181</td>\n",
       "      <td>180</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.552486</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>CCT|missense_variant|MODERATE|Gene_565549_5666...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18483</td>\n",
       "      <td>566361</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>172</td>\n",
       "      <td>171</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.581395</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>G|synonymous_variant|LOW|Gene_565549_566677|Ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18476</td>\n",
       "      <td>566244</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.167980e-15</td>\n",
       "      <td>171</td>\n",
       "      <td>170</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.584795</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>G|synonymous_variant|LOW|Gene_565549_566677|Ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6416</td>\n",
       "      <td>195428</td>\n",
       "      <td>CAAAAAAAAAAAAAAAGTAAAATAGAAAAGC</td>\n",
       "      <td>CAAAATAAGAAAAAAAGTAAAATAGAAAAGC,CAAAAAAAAAAAAA...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>24</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "      <td>1,1,1,6</td>\n",
       "      <td>37.500000</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>CAAAATAAGAAAAAAAGTAAAATAGAAAAGC|upstream_gene_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17103</td>\n",
       "      <td>528761</td>\n",
       "      <td>GTTTTTTTTTTTTTTTTTGAAAGA</td>\n",
       "      <td>GTTTTTTTTTTTTTTTTGAAAGA,GTTTTTTTTTTTTTTTTTTGAAAGA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>28</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>6,6</td>\n",
       "      <td>42.857143</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>GTTTTTTTTTTTTTTTTGAAAGA|upstream_gene_variant|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20563</td>\n",
       "      <td>629173</td>\n",
       "      <td>GTTTTTTTTTTTTTTTTAGTTTGAAC</td>\n",
       "      <td>GTTTTTTTTTATTTATAAGTTTGAAC,GTTTTTTTTTTTTTTTAGT...</td>\n",
       "      <td>1.383230e-14</td>\n",
       "      <td>30</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>1,2,6,2,2</td>\n",
       "      <td>43.333333</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>GTTTTTTTTTATTTATAAGTTTGAAC|upstream_gene_varia...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4753</td>\n",
       "      <td>141270</td>\n",
       "      <td>CAGAGAGAGAGAGAGAGAGAGC</td>\n",
       "      <td>CAGAGAGAGAGAGAGAGAGC,CAGAGAGAGAGAGAGAGAGAGAGC</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>52</td>\n",
       "      <td>27</td>\n",
       "      <td>25</td>\n",
       "      <td>7,18</td>\n",
       "      <td>48.076923</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>CAGAGAGAGAGAGAGAGAGC|upstream_gene_variant|MOD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19735</td>\n",
       "      <td>602077</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1.607630e+01</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>A|missense_variant|MODERATE|P02_orf1300|MPN489...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27073 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          POS                              REF  \\\n",
       "18482  566325                                C   \n",
       "18481  566311                                A   \n",
       "18485  566383                              GCC   \n",
       "18483  566361                                A   \n",
       "18476  566244                                A   \n",
       "...       ...                              ...   \n",
       "6416   195428  CAAAAAAAAAAAAAAAGTAAAATAGAAAAGC   \n",
       "17103  528761         GTTTTTTTTTTTTTTTTTGAAAGA   \n",
       "20563  629173       GTTTTTTTTTTTTTTTTAGTTTGAAC   \n",
       "4753   141270           CAGAGAGAGAGAGAGAGAGAGC   \n",
       "19735  602077                                T   \n",
       "\n",
       "                                                     ALT          QUAL  TOT  \\\n",
       "18482                                                  T  1.469330e-15  205   \n",
       "18481                                                  G  7.046470e-16  198   \n",
       "18485                                                CCT  0.000000e+00  181   \n",
       "18483                                                  G  0.000000e+00  172   \n",
       "18476                                                  G  1.167980e-15  171   \n",
       "...                                                  ...           ...  ...   \n",
       "6416   CAAAATAAGAAAAAAAGTAAAATAGAAAAGC,CAAAAAAAAAAAAA...  0.000000e+00   24   \n",
       "17103  GTTTTTTTTTTTTTTTTGAAAGA,GTTTTTTTTTTTTTTTTTTGAAAGA  0.000000e+00   28   \n",
       "20563  GTTTTTTTTTATTTATAAGTTTGAAC,GTTTTTTTTTTTTTTTAGT...  1.383230e-14   30   \n",
       "4753       CAGAGAGAGAGAGAGAGAGC,CAGAGAGAGAGAGAGAGAGAGAGC  0.000000e+00   52   \n",
       "19735                                                  A  1.607630e+01    1   \n",
       "\n",
       "       REFN  ALTN       ALTS        FRAC   SAMPLE  P  \\\n",
       "18482   204     1          1    0.487805  p2_2_S1  2   \n",
       "18481   197     1          1    0.505051  p2_2_S1  2   \n",
       "18485   180     1          1    0.552486  p2_2_S1  2   \n",
       "18483   171     1          1    0.581395  p2_2_S1  2   \n",
       "18476   170     1          1    0.584795  p2_2_S1  2   \n",
       "...     ...   ...        ...         ...      ... ..   \n",
       "6416     15     9    1,1,1,6   37.500000  p2_2_S1  2   \n",
       "17103    16    12        6,6   42.857143  p2_2_S1  2   \n",
       "20563    17    13  1,2,6,2,2   43.333333  p2_2_S1  2   \n",
       "4753     27    25       7,18   48.076923  p2_2_S1  2   \n",
       "19735     0     1          1  100.000000  p2_2_S1  2   \n",
       "\n",
       "                                                     EFF  \n",
       "18482  T|synonymous_variant|LOW|Gene_565549_566677|Ge...  \n",
       "18481  G|missense_variant|MODERATE|Gene_565549_566677...  \n",
       "18485  CCT|missense_variant|MODERATE|Gene_565549_5666...  \n",
       "18483  G|synonymous_variant|LOW|Gene_565549_566677|Ge...  \n",
       "18476  G|synonymous_variant|LOW|Gene_565549_566677|Ge...  \n",
       "...                                                  ...  \n",
       "6416   CAAAATAAGAAAAAAAGTAAAATAGAAAAGC|upstream_gene_...  \n",
       "17103  GTTTTTTTTTTTTTTTTGAAAGA|upstream_gene_variant|...  \n",
       "20563  GTTTTTTTTTATTTATAAGTTTGAAC|upstream_gene_varia...  \n",
       "4753   CAGAGAGAGAGAGAGAGAGC|upstream_gene_variant|MOD...  \n",
       "19735  A|missense_variant|MODERATE|P02_orf1300|MPN489...  \n",
       "\n",
       "[27073 rows x 12 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpcalls1[snpcalls1['SAMPLE']=='p2_2_S1'].sort_values('FRAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 'annotations' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-47-d335a4bfad19>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msimplify_effect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnpcalls1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenome\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-46-bf2fb86f217b>\u001b[0m in \u001b[0;36msimplify_effect\u001b[0;34m(df, positions, genome, modify_original, impact, mintot, minrefn, minaltn)\u001b[0m\n\u001b[1;32m      7\u001b[0m         \u001b[0mannotations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcorrespondance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenome\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgenome\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpositions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpositions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m         \u001b[0mannotations\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mannotations\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m     \u001b[0mannotations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mqualifiers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'label'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocation\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocation\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'-'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mannotations\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;34m'label'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mqualifiers\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m: local variable 'annotations' referenced before assignment"
     ]
    }
   ],
   "source": [
    "simplify_effect(snpcalls1, genome=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9c98a5e27e0e4195a0e684646f587140",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Figure\n",
    "\n",
    "\n",
    "def plot_analysis(df):\n",
    "    repro = [k for k, v in Counter(list(df.POS)).items() if v>=2]\n",
    "    plt.figure()\n",
    "    plt.subplot(3,3,1)\n",
    "    sns.lineplot(x='P', y='FRAC', data=df[(df['POS'].isin(posE)) & (df['POS'].isin(repro))], label='E')\n",
    "    sns.lineplot(x='P', y='FRAC', data=df[(df['POS'].isin(posN)) & (df['POS'].isin(repro))], label='NE')\n",
    "    sns.lineplot(x='P', y='FRAC', data=df[(df['POS'].isin(cas1)) & (df['POS'].isin(repro))], label='C')\n",
    "\n",
    "\n",
    "plot_analysis(snpcalls1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Substituion rate per base comparing cassette versus other distributions\n",
    "\n",
    "To evaluate the selection in the cassette we compare the rate of substitutions per base occuring within the cassette, a set of essential and non-essential genes and with the general distribution observed at genome level. In this case we do not care about the fraction each variant is found but how many times a variant is found in the cassette versus the other distributions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show the ratio of mutation in cassette versus general\n",
    "def mutation_rate(snpdf, segment, mintot=5, minrefn=3, minaltn=2):\n",
    "    _df = snpdf[(snpdf['TOT']>=mintot) & (snpdf['REFN']>=minrefn) & (snpdf['ALTN']>=minaltn)].copy()    \n",
    "    rs = {}\n",
    "    for sample in set(_df.SAMPLE):\n",
    "        total_len = 816394\n",
    "        _df2 = _df[_df['SAMPLE']==sample].copy()\n",
    "        NE = _df2[_df2['POS'].isin(posN)]\n",
    "        ES = _df2[_df2['POS'].isin(posE)]\n",
    "        TG = _df2[~_df2['POS'].isin(segment)]\n",
    "        CS = _df2[_df2['POS'].isin(segment)]\n",
    "        #total_len+=len(segment)\n",
    "        rs[sample] = [sample, sample.split('_')[0],\n",
    "                      int(sample.split('_')[0].replace('p', '').replace('3IPTG', '18')),\n",
    "                      TG.shape[0], ES.shape[0], NE.shape[0], CS.shape[0], \n",
    "                      sum(TG['FRAC']), sum(ES['FRAC']), sum(NE['FRAC']), sum(CS['FRAC'])]\n",
    "    jj = pd.DataFrame.from_dict(rs, orient='index')\n",
    "    jj.columns = ['SAMPLE', 'COND', 'P', 'T', 'E', 'N', 'C', 'TF', 'EF', 'NF', 'CF']\n",
    "    jj['T%'] = 100*jj['T']/(total_len)  \n",
    "    jj['E%'] = 100*jj['E']/len(posE)\n",
    "    jj['N%'] = 100*jj['N']/len(posN)\n",
    "    jj['C%'] = 100*jj['C']/len(segment)\n",
    "    return jj\n",
    "\n",
    "def tp(df, order):\n",
    "    rs = {}\n",
    "    c=0\n",
    "    for row, col in df.iterrows():\n",
    "        rs[c] = [col['COND'],'Essential', col['E%'], col['EF']]\n",
    "        rs[c+1] = [col['COND'],'Non-essential', col['N%'], col['NF']]\n",
    "        rs[c+2] = [col['COND'],'Chromosome', col['T%'], col['TF']]\n",
    "        rs[c+3] = [col['COND'],'Cassette', col['C%'], col['CF']]\n",
    "        c+=4\n",
    "    a = pd.DataFrame.from_dict(rs, orient='index')\n",
    "    a.columns = ['Condition', 'Loci', 'Variation per base [%]', 'Accumulated Fraction [%]']\n",
    "    return orderdf(a, ordered_classes=order, col='Condition')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutrate11 = mutation_rate(snpcalls1, segment=cas1)\n",
    "mutrate21 = mutation_rate(snpcalls2, segment=cas1)\n",
    "mutrate22 = mutation_rate(snpcalls2, segment=cas2)\n",
    "mutrate212 = mutation_rate(snpcalls2, segment=cas12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "25d6f7201093424d901273e1957b28ee",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.close('all')\n",
    "plt.figure(figsize=(15, 3))\n",
    "c = 1\n",
    "for text, df in zip(['C5 Par', 'C20 Par', 'C20 3b'], [mutrate11, mutrate21, mutrate22]):\n",
    "    plt.subplot(1,3,c)\n",
    "    if text=='C5 Par':\n",
    "        sns.barplot(x='Condition', y='Variation per base [%]', hue='Loci', palette='mako', data=tp(df, ['p2','p3','p15', 'p3IPTG']))\n",
    "    else:\n",
    "        sns.barplot(x='Condition', y='Variation per base [%]', hue='Loci', palette='mako', data=tp(df, ['p2','p3','p15']))\n",
    "    plt.ylim(0, 15)\n",
    "    c+=1\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dnds_functions import evolution_of_sample, dnds\n",
    "ttt = evolution_of_sample(effect1[effect1['IMPACT']=='HIGH'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "effectC= simplify_effect(snpcalls1, cas1, minaltn=3)\n",
    "effectE= simplify_effect(snpcalls1, posE, minaltn=3)\n",
    "effectN= simplify_effect(snpcalls1, posN, minaltn=3)\n",
    "\n",
    "tttC = evolution_of_sample(effectC)\n",
    "tttE = evolution_of_sample(effectE)\n",
    "tttN = evolution_of_sample(effectN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'p15_F_S1': [0.015,\n",
       "  0.011,\n",
       "  1.3636363636363638,\n",
       "  0.015,\n",
       "  0.011,\n",
       "  1.3636363636363638],\n",
       " 'p3IPTG_A_S1': [0.418,\n",
       "  0.265,\n",
       "  1.5773584905660376,\n",
       "  0.612,\n",
       "  0.327,\n",
       "  1.8715596330275228],\n",
       " 'p3_D_S3': [0.012, 0.0, '+', 0.012, 0.0, '+'],\n",
       " 'p2_2_S1': [0, 0, '/', 0, 0, '/'],\n",
       " 'p3IPTG_A_S1b': [0, 0, '/', 0, 0, '/'],\n",
       " 'p15_F_S5': [0, 0, '/', 0, 0, '/'],\n",
       " 'p2_2_S3': [0.009000000000000001,\n",
       "  0.022,\n",
       "  0.40909090909090917,\n",
       "  0.009000000000000001,\n",
       "  0.022,\n",
       "  0.40909090909090917],\n",
       " 'p2_2_S2': [0.006, 0.0, '+', 0.006, 0.0, '+'],\n",
       " 'p3_D_S2': [0.03,\n",
       "  0.06599999999999999,\n",
       "  0.4545454545454546,\n",
       "  0.03,\n",
       "  0.06599999999999999,\n",
       "  0.4545454545454546]}"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tttE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "155ba62f89ef40138481c9a168d64db1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x7f92ec6b1be0>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plt.figure()\n",
    "x, y = [[], []]\n",
    "for k, v in tttC.items():\n",
    "    if type(v[-1])==float and type(tttE[k][-1])==float:\n",
    "        x.append(tttE[k][-1])\n",
    "        y.append(v[-1])\n",
    "plt.scatter(x, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure()\n",
    "plt.subplot(2,2,1)\n",
    "sns.countplot(x='SAMPLE', data=snpcalls1)\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.title('Variant count per sample')\n",
    "\n",
    "plt.subplot(2,2,2)\n",
    "sns.boxplot(x='SAMPLE', y='FRAC', data=snpcalls1)\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.title('Fraction distribution per sample')\n",
    "\n",
    "plt.subplot(2,2,3)\n",
    "sns.countplot(x='SAMPLE', data=snpcalls2)\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.title('Variant count per sample')\n",
    "\n",
    "plt.subplot(2,2,4)\n",
    "sns.boxplot(x='SAMPLE', y='FRAC', data=snpcalls2)\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.title('Fraction distribution per sample')\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "General exploration of the most representative variants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "plt.figure(figsize=(15, 10))\n",
    "plt.subplot(3,1,1)\n",
    "sns.scatterplot(x='POS', y='FRAC', hue='SAMPLE', data=snpcalls1[(snpcalls1['FRAC']>=5) & (snpcalls1['TOT']>=10) & (snpcalls1['P']<=15) & (snpcalls1['FRAC']<100)])\n",
    "plt.axvspan(565510, 572114, color='gray', alpha=0.2)\n",
    "plt.xlabel('')\n",
    "plt.ylabel('Fraction of variant reads [%]')\n",
    "plt.xlim(1, 890000)\n",
    "\n",
    "plt.subplot(3,1,2)\n",
    "sns.scatterplot(x='POS', y='FRAC', hue='SAMPLE', data=snpcalls1[(snpcalls1['FRAC']>=5) & (snpcalls1['TOT']>=10) & (snpcalls1['P']>15) & (snpcalls1['FRAC']<100)])\n",
    "plt.axvspan(565510, 572114, color='gray', alpha=0.2)\n",
    "plt.xlabel('')\n",
    "plt.ylabel('Fraction of variant reads [%]')\n",
    "plt.xlim(1, 890000)\n",
    "\n",
    "plt.subplot(3,1,3)\n",
    "sns.scatterplot(x='POS', y='FRAC', hue='SAMPLE', data=snpcalls2[(snpcalls2['FRAC']>=5) & (snpcalls2['TOT']>=10) & (snpcalls2['FRAC']<100)])\n",
    "plt.axvspan(cass_par[0], cass_par[1], color='gray', alpha=0.2)\n",
    "plt.axvspan(cass_3b[0], cass_3b[1], color='gray', alpha=0.2)\n",
    "plt.xlabel('genome position [bp]')\n",
    "plt.ylabel('Fraction of variant reads [%]')\n",
    "plt.xlim(1, 890000)\n",
    "\n",
    "plt.savefig('./figures/supfig1.svg')\n",
    "plt.savefig('./figures/supfig1.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Selection of variants\n",
    "In this case we evaluate the fraction of each variant and how it is fixed along passages. As not all the samples are directly related, we consider the induced sample (IPTG) as 18. In this sense, if a mutation found in passages 2, 3 or 15 is found in 18 with a higher rate it implies that this variants are able to confere the capability to surpass the killswitch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 610,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_evolution(df):\n",
    "    positions = set(df[df['FRAC']>1].POS)\n",
    "    plt.figure()\n",
    "    plt.subplot(1,2,1)\n",
    "    sns.lineplot(x='PASS', y='FRAC', hue='POS', data=df[df['POS'].isin(positions)])\n",
    "    plt.subplot(1,2,2)\n",
    "    sns.boxplot(x='PASS', y='FRAC', data=df[df['POS'].isin(positions)])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 611,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/smiravet/.local/lib/python3.6/site-packages/ipykernel_launcher.py:3: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5d261d49ee3c4089a218c1c465670933",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-611-effa08ace5dc>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mplot_evolution\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-610-c53d1b5b5f59>\u001b[0m in \u001b[0;36mplot_evolution\u001b[0;34m(df)\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfigure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m     \u001b[0msns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlineplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'PASS'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'FRAC'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhue\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'POS'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'POS'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpositions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m     \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0msns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mboxplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'PASS'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'FRAC'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'POS'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpositions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/dist-packages/seaborn/relational.py\u001b[0m in \u001b[0;36mlineplot\u001b[0;34m(x, y, hue, size, style, data, palette, hue_order, hue_norm, sizes, size_order, size_norm, dashes, markers, style_order, units, estimator, ci, n_boot, sort, err_style, err_kws, legend, ax, **kwargs)\u001b[0m\n\u001b[1;32m   1082\u001b[0m         \u001b[0max\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgca\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1083\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1084\u001b[0;31m     \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1085\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1086\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/dist-packages/seaborn/relational.py\u001b[0m in \u001b[0;36mplot\u001b[0;34m(self, ax, kws)\u001b[0m\n\u001b[1;32m    764\u001b[0m         \u001b[0;31m# Loop over the semantic subsets and draw a line for each\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    765\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 766\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0msemantics\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubset_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    767\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    768\u001b[0m             \u001b[0mhue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msize\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstyle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msemantics\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/dist-packages/seaborn/relational.py\u001b[0m in \u001b[0;36msubset_data\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    330\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    331\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 332\u001b[0;31m                 \u001b[0msubset_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msort_df\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubset_data\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"units\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"x\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"y\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    333\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    334\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munits\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/dist-packages/seaborn/utils.py\u001b[0m in \u001b[0;36msort_df\u001b[0;34m(df, *args, **kwargs)\u001b[0m\n\u001b[1;32m     43\u001b[0m     \u001b[0;34m\"\"\"Wrapper to handle different pandas sorting API pre/post 0.17.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     44\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 45\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     46\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/dist-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36msort_values\u001b[0;34m(self, by, axis, ascending, inplace, kind, na_position)\u001b[0m\n\u001b[1;32m   5001\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5002\u001b[0m         new_data = self._data.take(\n\u001b[0;32m-> 5003\u001b[0;31m             \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_block_manager_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverify\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5004\u001b[0m         )\n\u001b[1;32m   5005\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/dist-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mtake\u001b[0;34m(self, indexer, axis, verify, convert)\u001b[0m\n\u001b[1;32m   1395\u001b[0m         \u001b[0mnew_labels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1396\u001b[0m         return self.reindex_indexer(\n\u001b[0;32m-> 1397\u001b[0;31m             \u001b[0mnew_axis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnew_labels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallow_dups\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1398\u001b[0m         )\n\u001b[1;32m   1399\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/dist-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mreindex_indexer\u001b[0;34m(self, new_axis, indexer, axis, fill_value, allow_dups, copy)\u001b[0m\n\u001b[1;32m   1265\u001b[0m                     ),\n\u001b[1;32m   1266\u001b[0m                 )\n\u001b[0;32m-> 1267\u001b[0;31m                 \u001b[0;32mfor\u001b[0m \u001b[0mblk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1268\u001b[0m             ]\n\u001b[1;32m   1269\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/dist-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m   1265\u001b[0m                     ),\n\u001b[1;32m   1266\u001b[0m                 )\n\u001b[0;32m-> 1267\u001b[0;31m                 \u001b[0;32mfor\u001b[0m \u001b[0mblk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1268\u001b[0m             ]\n\u001b[1;32m   1269\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/dist-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mtake_nd\u001b[0;34m(self, indexer, axis, new_mgr_locs, fill_tuple)\u001b[0m\n\u001b[1;32m   1312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1313\u001b[0m         new_values = algos.take_nd(\n\u001b[0;32m-> 1314\u001b[0;31m             \u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallow_fill\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mallow_fill\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfill_value\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfill_value\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1315\u001b[0m         )\n\u001b[1;32m   1316\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/dist-packages/pandas/core/algorithms.py\u001b[0m in \u001b[0;36mtake_nd\u001b[0;34m(arr, indexer, axis, out, fill_value, mask_info, allow_fill)\u001b[0m\n\u001b[1;32m   1649\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mis_extension_array_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1650\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfill_value\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfill_value\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallow_fill\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mallow_fill\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1651\u001b[0;31m     \u001b[0;32melif\u001b[0m \u001b[0mis_datetime64tz_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1652\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfill_value\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfill_value\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallow_fill\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mallow_fill\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1653\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mis_interval_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "plot_evolution(data1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'LacI4': [565549, 566678, '-'],\n",
       " 'pS': [566678, 566721, '-'],\n",
       " 'Par cassette': [566747, 567674, '-'],\n",
       " 'cas9B': [567711, 571818, '-'],\n",
       " 'pG64': [571818, 571934, '-'],\n",
       " 'gRNA2': [571942, 572044, '-'],\n",
       " 'protospacer12 (10 targets)': [572024, 572044, '-'],\n",
       " 'p438': [572044, 572066, '-'],\n",
       " 'IR-OR': [572088, 572114, '-'],\n",
       " 'MTn insertion point': [572113, 572124, '-']}"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "376.3333333333333\n",
      "1369.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/smiravet/.local/lib/python3.6/site-packages/Bio/Seq.py:2309: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning)\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "genes_affected = {}\n",
    "for k, v in annotations.items():\n",
    "    if k in ['cas9B', 'LacI4']:\n",
    "        if k=='cas9B':\n",
    "            seq = Seq(genome_C5[v[0]:v[1]])\n",
    "        else:\n",
    "            seq = Seq(genome_C5[v[0]:v[1]])\n",
    "        print(len(seq)/3)\n",
    "        rvc = seq.reverse_complement()\n",
    "        mrn = rvc.transcribe()\n",
    "        prt = mrn.translate(table=4)\n",
    "        genes_affected[k] = [seq, rvc, mrn, prt, len(prt)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('RLESGQ*R*V', HasStopCodon(ExtendedIUPACProtein(), '*'))"
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_affected['LacI4'][3][-10:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "LacI4 has extra bases in the annotation, that makes it to be non-codonic. I will consider the first stop found (move the end 10 bases)..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'annotations' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-33-84fcf2bf1dbc>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSeq\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mSeq\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mgenes_affected\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mannotations\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'cas9B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'LacI4'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'cas9B'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'annotations' is not defined"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "genes_affected = {}\n",
    "for k, v in annotations.items():\n",
    "    if k in ['cas9B', 'LacI4']:\n",
    "        if k=='cas9B':\n",
    "            seq = Seq(genome_seq[v[0]:v[1]])\n",
    "        else:\n",
    "            seq = Seq(genome_seq[v[0]+10:v[1]])\n",
    "        print(len(seq)/3)\n",
    "        rvc = seq.reverse_complement()\n",
    "        mrn = rvc.transcribe()\n",
    "        prt = mrn.translate(table=4)\n",
    "        genes_affected[k] = [seq, rvc, mrn, prt, len(prt)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'LacI4'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-34-2cf0e14076cd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mgenes_affected\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'LacI4'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m: 'LacI4'"
     ]
    }
   ],
   "source": [
    "genes_affected['LacI4'][3][-10:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Evaluate effect of the selected mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>P</th>\n",
       "      <th>EFF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4.927350e-15</td>\n",
       "      <td>60</td>\n",
       "      <td>58</td>\n",
       "      <td>2</td>\n",
       "      <td>3.333333</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>149</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>75</td>\n",
       "      <td>74</td>\n",
       "      <td>1</td>\n",
       "      <td>1.333333</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>150</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>2.263910e-15</td>\n",
       "      <td>69</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>1.449275</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>154</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>62</td>\n",
       "      <td>61</td>\n",
       "      <td>1</td>\n",
       "      <td>1.612903</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>176</td>\n",
       "      <td>TAAT</td>\n",
       "      <td>CAAC</td>\n",
       "      <td>3.420820e-15</td>\n",
       "      <td>58</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>1.724138</td>\n",
       "      <td>p2_2_S1</td>\n",
       "      <td>2</td>\n",
       "      <td>CAAC|upstream_gene_variant|MODIFIER|dnaN|MPN00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41166</td>\n",
       "      <td>822775</td>\n",
       "      <td>AACGT</td>\n",
       "      <td>GACGC</td>\n",
       "      <td>3.631530e-15</td>\n",
       "      <td>171</td>\n",
       "      <td>170</td>\n",
       "      <td>1</td>\n",
       "      <td>0.584795</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>GACGC|missense_variant|MODERATE|soj|MPN688|tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41167</td>\n",
       "      <td>822787</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.152910e-15</td>\n",
       "      <td>175</td>\n",
       "      <td>173</td>\n",
       "      <td>2</td>\n",
       "      <td>1.142857</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>G|synonymous_variant|LOW|soj|MPN688|transcript...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41168</td>\n",
       "      <td>822830</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>3.857310e-15</td>\n",
       "      <td>165</td>\n",
       "      <td>164</td>\n",
       "      <td>1</td>\n",
       "      <td>0.606061</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>G|missense_variant|MODERATE|soj|MPN688|transcr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41169</td>\n",
       "      <td>822867</td>\n",
       "      <td>TTTTT</td>\n",
       "      <td>CTTTC</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>172</td>\n",
       "      <td>171</td>\n",
       "      <td>1</td>\n",
       "      <td>0.581395</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>CTTTC|missense_variant|MODERATE|soj|MPN688|tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41170</td>\n",
       "      <td>822913</td>\n",
       "      <td>CGCGCGTAT</td>\n",
       "      <td>TGCGTGAAA</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>161</td>\n",
       "      <td>160</td>\n",
       "      <td>1</td>\n",
       "      <td>0.621118</td>\n",
       "      <td>p3IPTG_A_S1b</td>\n",
       "      <td>18</td>\n",
       "      <td>TGCGTGAAA|upstream_gene_variant|MODIFIER|K05_o...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>508440 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          POS        REF        ALT          QUAL  TOT  REFN  ALTN      FRAC  \\\n",
       "0          70          C          T  4.927350e-15   60    58     2  3.333333   \n",
       "1         149          T          C  0.000000e+00   75    74     1  1.333333   \n",
       "2         150          A          T  2.263910e-15   69    68     1  1.449275   \n",
       "3         154          T          C  0.000000e+00   62    61     1  1.612903   \n",
       "4         176       TAAT       CAAC  3.420820e-15   58    57     1  1.724138   \n",
       "...       ...        ...        ...           ...  ...   ...   ...       ...   \n",
       "41166  822775      AACGT      GACGC  3.631530e-15  171   170     1  0.584795   \n",
       "41167  822787          A          G  1.152910e-15  175   173     2  1.142857   \n",
       "41168  822830          T          G  3.857310e-15  165   164     1  0.606061   \n",
       "41169  822867      TTTTT      CTTTC  0.000000e+00  172   171     1  0.581395   \n",
       "41170  822913  CGCGCGTAT  TGCGTGAAA  0.000000e+00  161   160     1  0.621118   \n",
       "\n",
       "             SAMPLE   P                                                EFF  \n",
       "0           p2_2_S1   2  T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "1           p2_2_S1   2  C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "2           p2_2_S1   2  T|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "3           p2_2_S1   2  C|upstream_gene_variant|MODIFIER|dnaN|MPN001|t...  \n",
       "4           p2_2_S1   2  CAAC|upstream_gene_variant|MODIFIER|dnaN|MPN00...  \n",
       "...             ...  ..                                                ...  \n",
       "41166  p3IPTG_A_S1b  18  GACGC|missense_variant|MODERATE|soj|MPN688|tra...  \n",
       "41167  p3IPTG_A_S1b  18  G|synonymous_variant|LOW|soj|MPN688|transcript...  \n",
       "41168  p3IPTG_A_S1b  18  G|missense_variant|MODERATE|soj|MPN688|transcr...  \n",
       "41169  p3IPTG_A_S1b  18  CTTTC|missense_variant|MODERATE|soj|MPN688|tra...  \n",
       "41170  p3IPTG_A_S1b  18  TGCGTGAAA|upstream_gene_variant|MODIFIER|K05_o...  \n",
       "\n",
       "[508440 rows x 11 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snpcalls1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-36-96ce78c8031b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mrs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mselected\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miterrows\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0meff\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mEFF\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m','\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0mjjj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0meff\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'|'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'selected' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-184-19ff2ad0f5a8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mselected\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'selected' is not defined"
     ]
    }
   ],
   "source": [
    "selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected_ann' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-38-28fde1192669>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mselected_ann\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'selected_ann' is not defined"
     ]
    }
   ],
   "source": [
    "selected_ann"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected_ann' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-39-e70eb41dc56f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mselected_ann\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mselected_ann\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'IMPACT'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'HIGH'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'selected_ann' is not defined"
     ]
    }
   ],
   "source": [
    "selected_ann[selected_ann['IMPACT']=='HIGH']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation rate study\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.close('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'mutrate1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-44-d63c0ec7e1ed>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmelt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutrate1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'P'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmutrate1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;34m'%'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'mutrate1' is not defined"
     ]
    }
   ],
   "source": [
    "pd.melt(mutrate1.sort_values('P')[[i for i in mutrate1.columns if '%' in i]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d0f22cbb724f46aebea1cae18e8be94f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "NameError",
     "evalue": "name 'mutrate1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-45-05ad25cb1d90>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfigure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mboxplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmutrate1\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'N%'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutrate2\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'N%'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'mutrate1' is not defined"
     ]
    }
   ],
   "source": [
    "plt.figure()\n",
    "plt.boxplot([mutrate1['N%'], mutrate2['N%']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d0774f1234ea48ddac4ae853136b8b18",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "NameError",
     "evalue": "name 'mutrate1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-46-c03ce5ec3171>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'OUT%'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IN%'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m16\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m16\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'grey'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlinestyle\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'--'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0msns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscatterplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmutrate1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutrate1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     plt.text(x=mutrate1[x][i]+0.2,y=mutrate1[y][i]+0.3,s=mutrate1.index[i],\n",
      "\u001b[0;31mNameError\u001b[0m: name 'mutrate1' is not defined"
     ]
    }
   ],
   "source": [
    "plt.close('all')\n",
    "plt.figure()\n",
    "x, y = ['OUT%', 'IN%']\n",
    "plt.plot([0,16],[0,16], c='grey', linestyle='--')\n",
    "sns.scatterplot(x=x, y=y, data=mutrate1)\n",
    "for i in range(mutrate1.shape[0]):\n",
    "    plt.text(x=mutrate1[x][i]+0.2,y=mutrate1[y][i]+0.3,s=mutrate1.index[i],\n",
    "             fontdict=dict(size=8))\n",
    "sns.scatterplot(x=x, y=y, data=mutrate2)\n",
    "c=0\n",
    "for i in range(mutrate2.shape[0]):\n",
    "\n",
    "    if c==0:\n",
    "        plt.text(x=mutrate2[x][i]+0.2,y=mutrate2[y][i]+0.4,s=mutrate2.index[i],\n",
    "                 fontdict=dict(size=10))\n",
    "        c==1\n",
    "    else:\n",
    "        plt.text(x=mutrate2[x][i]+0.2,y=mutrate2[y][i]+0.2,s=mutrate2.index[i],\n",
    "                 fontdict=dict(size=8))\n",
    "        c==0\n",
    "plt.xlabel('Genome Variant Rate [%]')\n",
    "plt.ylabel('Cassette Variant Rate [%]')\n",
    "plt.xlim(0,16)\n",
    "plt.ylim(0,16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'mutrate' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-47-8c5c8001c3c9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mwrs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mtrs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0mns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mCounter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutrate\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mP\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutrate\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mP\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     wrs[p] = wilcoxon(x=list(mutrate[mutrate['P']==p]['OUT%']),\n",
      "\u001b[0;31mNameError\u001b[0m: name 'mutrate' is not defined"
     ]
    }
   ],
   "source": [
    "from scipy.stats import wilcoxon, ttest_rel\n",
    "from collections import Counter\n",
    "\n",
    "wrs = {}\n",
    "trs = {}\n",
    "ns = dict(Counter(mutrate.P))\n",
    "for p in set(mutrate.P):\n",
    "    wrs[p] = wilcoxon(x=list(mutrate[mutrate['P']==p]['OUT%']),\n",
    "                      y=list(mutrate[mutrate['P']==p]['IN%']))[1]\n",
    "    trs[p] = ttest_rel(list(mutrate[mutrate['P']==p]['OUT%']),\n",
    "                       list(mutrate[mutrate['P']==p]['IN%']))[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ns' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-48-5ec18e92d2c6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwrs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtrs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'ns' is not defined"
     ]
    }
   ],
   "source": [
    "ns, wrs, trs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "null hypothesis (same mut rate) cannot be rejected at a confidence level of 5%, only at early passages 2 and 3 seems to be limited"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'snpcalls50' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-49-bf7c51358a6e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msnpcalls50\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'snpcalls50' is not defined"
     ]
    }
   ],
   "source": [
    "snpcalls50"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot variations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[565510, 572124]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_par"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-51-19ff2ad0f5a8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mselected\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'selected' is not defined"
     ]
    }
   ],
   "source": [
    "selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from reportlab.lib import colors\n",
    "from reportlab.lib.units import cm\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "\n",
    "\n",
    "def plot_variations(genbank_file, coords, variations):\n",
    "    record = SeqIO.read(genbank_file, \"genbank\")\n",
    "\n",
    "    gd_diagram = GenomeDiagram.Diagram(record.id)\n",
    "    gd_track_for_features = gd_diagram.new_track(1, name=\"Annotated Features\")\n",
    "    gd_feature_set = gd_track_for_features.new_set()\n",
    "\n",
    "    for feature in record.features:\n",
    "        if feature.location.start>=coords[0] and feature.location.start<=coords[1] and feature.type!='Polymorphism':\n",
    "            print(feature.type, feature.qualifiers['label'], feature.location)\n",
    "            feature.location.strand=-1\n",
    "            #if feature.type != \"gene\":\n",
    "                # Exclude this feature\n",
    "            #    continue\n",
    "            #if len(gd_feature_set) % 2 == 0:\n",
    "            #    color = colors.blue\n",
    "            if feature.type=='CDS':\n",
    "                color = colors.lightblue\n",
    "                gd_feature_set.add_feature(\n",
    "                    feature, sigil=\"ARROW\", arrowshaft_height=1.0, color=color, height=0.2, label=True, label_size=14, label_angle=0\n",
    "                )\n",
    "            # else:\n",
    "            #     color = colors.lightblue\n",
    "            #     gd_feature_set.add_feature(\n",
    "            #         feature, sigil=\"BOX\", color=color, height=0.1, label=True, label_size=14, label_angle=10)\n",
    "    colordic = {'stop': colors.firebrick, 'frameshift':colors.red, 'missense':colors.orange}\n",
    "    done = []\n",
    "\n",
    "    for pos, eff in zip(variations.POS, variations.EFF):\n",
    "        if pos not in done:\n",
    "            effi = eff.split('|')[1].split('_')[0]\n",
    "            color = colordic[effi]\n",
    "            feature = SeqFeature(FeatureLocation(pos, pos+1))\n",
    "            gd_feature_set.add_feature(\n",
    "                    feature,\n",
    "                    color=color,\n",
    "                    name=' '+effi,\n",
    "                    label=True,\n",
    "                    label_size=15,\n",
    "                    size=20\n",
    "                    #label_color=color,\n",
    "                )\n",
    "            done.append(pos)\n",
    "            for i in range(pos-30, pos+30):\n",
    "                done.append(i)\n",
    "    gd_diagram.draw(format=\"linear\", pagesize=\"A5\", fragments=1, start=565510, end=572124)\n",
    "    gd_diagram.write(\"plasmid_linear_nice.svg\", \"SVG\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'selected' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-53-3c5b915f50d3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mplot_variations\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'./data/C5_refSeq.gb'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcoords\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m565549\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m571818\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariations\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mselected\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'selected' is not defined"
     ]
    }
   ],
   "source": [
    "plot_variations('./data/C5_refSeq.gb', coords=[565549,571818], variations=selected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 653,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 637,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "268.6666666666668\n"
     ]
    }
   ],
   "source": [
    "raw_seq, new_seq = get_new_sequence(seq_record5, 'LacI4', pos=[565576, 'AGGC', 'GGGT'])\n",
    "yyy = dnds(raw_seq, new_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 654,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 648,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>PASS</th>\n",
       "      <th>POS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>TOT</th>\n",
       "      <th>REFN</th>\n",
       "      <th>ALTN</th>\n",
       "      <th>FRAC</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFF</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>AFF</th>\n",
       "      <th>MUT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2140</td>\n",
       "      <td>p3_D_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>571815</td>\n",
       "      <td>0.0</td>\n",
       "      <td>674</td>\n",
       "      <td>668</td>\n",
       "      <td>6</td>\n",
       "      <td>0.898204</td>\n",
       "      <td>CCAGATGTAA</td>\n",
       "      <td>TCTGTTGTTT</td>\n",
       "      <td>start_lost</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>cas9B</td>\n",
       "      <td>p.LeuAsp1?</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       SAMPLE  PASS     POS  QUAL  TOT  REFN  ALTN      FRAC         REF  \\\n",
       "2140  p3_D_S2     3  571815   0.0  674   668     6  0.898204  CCAGATGTAA   \n",
       "\n",
       "             ALT         EFF IMPACT    AFF         MUT  \n",
       "2140  TCTGTTGTTT  start_lost   HIGH  cas9B  p.LeuAsp1?  "
      ]
     },
     "execution_count": 648,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effect1[effect1['POS']==571815]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 656,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "p3IPTG_A_S1 2.404754044239023\n",
      "p3_D_S2 1.7427616926503313\n",
      "p15_F_S1 2.3935483870967738\n",
      "p15_F_S5 3.4130506790327924\n",
      "p3IPTG_A_S1b 2.6608581600149894\n",
      "p3_D_S3 2.117820324005891\n",
      "p2_2_S3 2.3640178337267255\n",
      "p2_2_S1 2.127371273712737\n",
      "p2_2_S2 1.7000000000000002\n"
     ]
    }
   ],
   "source": [
    "for k, v in ttt.items():\n",
    "    print(k, v[-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 620,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-56-9a8f373c509c>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-56-9a8f373c509c>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    df = pd.read_csv(, header=60, sep='\\t')\u001b[0m\n\u001b[0m                     ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(, header=60, sep='\\t')\n",
    "df2 = df[(df['INFO'].str.contains('CIGAR=1X;'))].sample(n=150, random_state=1).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "Missing parentheses in call to 'print'. Did you mean print(args.query, chromosome, pn, ps, pns, dn, ds, dns)? (<ipython-input-57-012244273d6f>, line 18)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-57-012244273d6f>\"\u001b[0;36m, line \u001b[0;32m18\u001b[0m\n\u001b[0;31m    print args.query, chromosome, pn, ps, pns, dn, ds, dns\u001b[0m\n\u001b[0m             ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m Missing parentheses in call to 'print'. Did you mean print(args.query, chromosome, pn, ps, pns, dn, ds, dns)?\n"
     ]
    }
   ],
   "source": [
    "pn, ps, pns, dn, ds, dns = 0, 0, 0, 0, 0, 0\n",
    "for item in sorted(output[chromosome]):\n",
    "    raw_seq, new_seq = get_new_sequence(args.reference, chromosome, output[chromosome][item][0].qualifiers[\n",
    "        'locus_tag'][0], output[chromosome][item][1])\n",
    "    newpn, newps, newdn, newds = dnds(raw_seq, new_seq)\n",
    "    pn += newpn\n",
    "    ps += newps\n",
    "    dn += newdn\n",
    "    ds += newds\n",
    "if pn == 0 and ps == 0:\n",
    "    pns, dns = '/', '/'\n",
    "elif ps == 0:\n",
    "    pns, dns = '+', '+'\n",
    "elif pn == 0:\n",
    "    pns, dns = '-', '-'\n",
    "else:\n",
    "    pns, dns = pn / ps, dn / ds\n",
    "print args.query, chromosome, pn, ps, pns, dn, ds, dns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-58-9fd2afaa1715>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-58-9fd2afaa1715>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    for records in SeqIO.parse(, \"genbank\"):\u001b[0m\n\u001b[0m                               ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "for records in SeqIO.parse(, \"genbank\"):\n",
    "    print(/)\n",
    "    raw_seq, new_seq = get_new_sequence('./data/C5_refSeq.gb', 'C5', output[chromosome][item][0].qualifiers[\n",
    "        'locus_tag'][0], output[chromosome][item][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  },
  "toc-showcode": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
